---
title: "PD, microbiome and host genetics on SNCA and HLA variants"
subtitle: "Data: NGRC"
author: "Giacomo Antonello, Zachary D. Wallen, Charles F. Murchinson, Levi Waldron, Haydeh Payami"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
---

```{r, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE
)

default_outdir <- "output_NGRC/"

dir.create(default_outdir, showWarnings = FALSE)
dir.create(file.path(default_outdir, "Q1"))
dir.create(file.path(default_outdir, "Q2"))
dir.create(file.path(default_outdir, "Q3"))

```

```{r, include=FALSE}

# Check and install necessary CRAN packages
CRAN_required_packages <- c("tidyverse", "gwasrapidd", "DT", "kableExtra", "data.table", "ggpubr", "BiocManager", "readxl","kableExtra", "plotly", "modeest")

install.packages(
  CRAN_required_packages[!(CRAN_required_packages %in% installed.packages())] 
)

# Check and install necessary Bioconductor packages
BioC_required_packages <- c("Maaslin2", "ANCOMBC", "biomaRt", "mia", "phyloseq", "microbiome","SIAMCAT", "fgsea")
BioC_required_packages <- BioC_required_packages[!(BioC_required_packages %in% installed.packages())]

if(length(BioC_required_packages) > 0) {
  BiocManager::install(BioC_required_packages[!(BioC_required_packages %in% installed.packages())],
                       ask = FALSE,
                       update = TRUE)
}


# load basic data manipulation packages
library(DT)
library(kableExtra)
library(MultiAssayExperiment)
library(ggpubr)
library(tidyverse)

theme_set(theme_light())
# load previously wrangled data
load("../Data/NGRC_data.mae.RData")
  
rm(BioC_required_packages,CRAN_required_packages)
gc(verbose = FALSE)
```

Below is the data sources for Wallen et al. 2022 (Nat Comms). Data have been 
further prepared by Giacomo to make one single data object that would hopefully, 
speed up the preparation phase. This data object contains microbiome, pathway, 
participants data and genetic data.

```{r}
NGRC_data.mae
```

Genetic data have been retrieved with the protocol available in `Data/genetic_data_preparation.{Rmd,html}`.
In brief:

  1. The GWAS catalog was searched for variants associated with the ontology term
  `MONDO_0005180`, and downloaded. all this with 
  [gwasrapidd](https://rmagno.eu/gwasrapidd/articles/gwasrapidd.html). 
  Catalog's version is not stated, I'm assuming that since the packages uses the `REST API`, 
  it should be the newest.
  
  2. Gene annotation was taken from `biomaRt`, but the GWAS catalog had annotations
  too. Naming is based on the HUGO nomenclature.

# Approach 1 - Sum bacteria in each cluster together

## Define `ClusterSum`

Haydeh has manually curated a list of bacteria that go together into 6 different
clusters elevated or decreased in PD. To see the genetic impact of PD associated
variants on these bacteria without having to pay a high multiple testing price.

Our first approach is to do the following:

  1. Define which species belong to which cluster
  2. Divide % abundances by 100, to obtain relative abundances again
  3. Sum relative abundances of all species within each cluster

```{r}
biomarkers_medrxiv.df <- readxl::read_xlsx("../Data/Biomarkers_Payami_medRxiv.xlsx")

datatable(biomarkers_medrxiv.df %>% select(-species_metaphlan, `Number of connections`), caption = "List of Bacterial species and their cluster assignment as manually curated by prof. Haydeh Payami")
```

```{r}

cluster_sums.df <- left_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(NGRC_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  split.data.frame(f = .$cluster_name) %>% 
  lapply(select, all_of(MultiAssayExperiment::colnames(NGRC_data.mae)[["metaphlan_RelAbund.se"]])) %>% 
  lapply(colSums) %>% 
  bind_cols(SubjectID = names(.[[1]])) %>% 
  relocate(SubjectID)

cluster_sums.df %>% 
  mutate_if(is.numeric, round, 3) %>% 
datatable(caption = "Snippet of values obtained with the sum method across clusters. These will be the variables we will use to test association with variants' dosages")
```

For each cluster, see how many species were correctly mapped to the NGRC dataset.

```{r}

n_all_species <- biomarkers_medrxiv.df %>% 
  count(cluster_name)

n_mapped_species <- inner_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(NGRC_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  count(cluster_name)

unmapped_species <- biomarkers_medrxiv.df %>%
  mutate(Species2 = case_when(!(species_metaphlan %in% rownames(
    assay(NGRC_data.mae, "metaphlan_RelAbund.se")
  )) ~ Species,
  TRUE ~ "")) %>% group_by(cluster_name) %>% reframe(unmapped = paste0(Species2[Species2 != ""], collapse = ", "))

full_join(n_all_species, n_mapped_species, by = "cluster_name") %>% 
  full_join(unmapped_species, by = "cluster_name") %>% 
  set_names(c("Cluster Name", "Total", "N. Mapped", "Not Mapped")) %>% 
  arrange(desc(Total)) %>% 
  kbl(caption = "Number of total species and correctly mapped species in the NGRC dataset")

```

These variables are far from normal, and log transformation does not help.

```{r, fig.cap= "Distribution of cumulative sum of relative abundances of cluster member species."}
cluster_sums.df %>% 
  select(-SubjectID) %>% 
  reshape2::melt() %>% 
  mutate(
    log2Transf = log2(value + 1),
    sq.root = sqrt(value)
    ) %>% 
  reshape2::melt(
    variable.name = "Transformation"
    ) %>% 
  mutate(
    Transformation = case_when(
      Transformation == "value" ~ 'untransformed',
      Transformation == "log2Transf" ~ 'log2',
      Transformation == "sq.root" ~ 'sq.root',
      TRUE ~ 'error'
      )
    ) %>% 
  ggplot(aes(x = value, color = Transformation)) + 
  geom_histogram(fill = "transparent") + 
  facet_wrap(~variable) 
```

## Define variants of interest and number of independent variants tested

Genetic variants were included if retained if:

  - they were associated with PD in any GWAS catalog entry (P < 5e-8)
  - they were found in the TOPMed imputed genotype data
  - they had a gene name associated to them (for easy lookup, but exception may
  be missed)

```{r}

samples_to_keep.df <- colData(NGRC_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() %>% 
  filter(
    !is.na(Age_at_collection),
    !is.na(seqs_scaled),
    !is.na(Case_status)
    )

variants_to_keep_rowData.df <- as.data.frame(rowData(NGRC_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(NGRC_data.mae[["genotypes_dosages.se"]])$GENE),])
rownames(variants_to_keep_rowData.df) <- variants_to_keep_rowData.df$RSID

per_sample_dosages.df <- t(assay(NGRC_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

data_alltogether.df <- inner_join(
  # clustered bacteria variables
  cluster_sums.df, 
  # variants available to test
  per_sample_dosages.df, by = "SubjectID") %>% 
  inner_join(
  # other individual data (metadata), excluding unwanted samples
    samples_to_keep.df, by = "SubjectID")

as_tibble(variants_to_keep_rowData.df) %>% 
  set_names(
    gsub("INFO_", "", colnames(.))
    ) %>%
  rename(
    `Empirical (R2)` = ER2,
    `Imputation Quality (R2)` = R2,
         ) %>% 
  datatable(caption = "List of variants found in the GWAS catalog that are explicitly mapped to the SNCA or HLA genes.")
```

```{r LDlink parameters setup}
r2_thresh <- 0.1
refPop <- "CEU"
maf_thresh <- 0.01
```

Some of these variants are in high LD. We want to adjust for the total number of 
independent tests, therefore we adjust for variants in low LD only across the 
genes of interest. We used an R API based on NIH's tool `LDlink`. Parameters
chosen are:

  - r2 threshold = `r r2_thresh`
  - reference population (1000 Genomes) = `r refPop`
  - MAF threshold = `r maf_thresh`

```{r}

cache(
  variants_pruning_step <- split.data.frame(variants_to_keep_rowData.df,variants_to_keep_rowData.df$CHROM) %>% 
  lapply( 
    function(x)
    LDlinkR::SNPclip(
      r2_threshold = r2_thresh,
      snps = x$RSID,
      pop = refPop,
      maf_threshold = maf_thresh,
      token = "78e50d5fd7e7", 
      genome_build = "grch38"
      )
  ), dir = file.path(default_outdir, "variant_pruning_cache")
  )

variants_pruning_table <- tryCatch(lapply(
  variants_pruning_step, mutate, Details = ifelse(grepl("removed", Details), "Removed", "Kept")
  ) %>% 
  lapply(group_by, Details) %>%
  lapply(tally) %>%
  bind_rows(.id = "Chromosome") %>% 
  pivot_wider(names_from = "Details", values_from = "n") %>% 
  mutate(Total = Kept + Removed),
  error = function(e) return(NA))

kbl(variants_pruning_table, caption = paste0("Number of idependent variants ('Kept') based on LDlink calculations on default parameters (R2 = ", r2_thresh, ", MAF = ", maf_thresh, "). The population chosen is 'CEU'. The sum of 'Kept' values represents the number of independent tests that need to be accounted for later on. NB: All variants will be tested, but the number of independent tests will be based on this 'Kept' column.")) %>% 
  kable_styling()
```

```{r define number of tests to account for}
n_indep_variants <- ifelse(all(is.na(variants_pruning_table)), 6, sum(sapply(variants_pruning_step, function(x) sum(grepl("kept", x$Details)))))
```

## Sample inclusion criteria and statistical modeling

  - Available data for age, sequencing depth, microbiome PD cumulative scores (N unchanged)
  
  - Collection method should not be `swab` (N goes from  `r nrow(cluster_sums.df)` to `r nrow(samples_to_keep.df)`)
  
  - Available genetic data (N goes from `r nrow(samples_to_keep.df)` to `r nrow(data_alltogether.df)`)

Below is some information of the variants associated with PD found in our 
imputed genotypes.

```{r}
cbind(data_alltogether.df %>% 
  select(starts_with("chr")) %>% 
  mutate_all(round, 0) %>% 
  sapply(table) %>% 
  t(), 
  data_alltogether.df %>% 
    select(starts_with("chr")) %>% 
    mutate_all(round, 0) %>% 
    sapply(function(x) table(x) %>% prop.table()) %>% 
    t()) %>% 
  .[,c(1,4,2,5,3,6)] %>%   
  as.data.frame() %>% 
  set_names(make.unique(colnames(.))) %>% 
  rownames_to_column("id") %>%
  separate_wider_delim(cols = id, names = c("CHROM", "POS", "REF", "ALT"), delim = "_", too_few = "align_start") %>% 
  rename(
    `0 (N)` = `0`,
    `0 (%)` = `0.1`,
    `1 (N)` = `1`,
    `1 (%)` = `1.1`,
    `2 (N)` = `2`,
    `2 (%)` = `2.1`
  ) %>% 
  mutate_at(.vars = vars(c(`0 (%)`,`1 (%)`,`2 (%)`)), .funs = function(x) round(x*100, 2)) %>% 
  
  kbl(caption = "Number of samples with alternate allele dosages. NB: no distinction is done for 1|0 or 0|1, as both carry one alternate allese. Values are shown as total counts and percentages") %>% 
  kable_styling() %>% 
  add_header_above(c("Variant Information" = 4, "Distribution of dosages in the cohort studied" = 6)) %>% 
  column_spec(c(4,6,8), border_right = TRUE)
```


## Q1 - Is there interaction between SNP variant dosage and PD status (case or control) on cluster abundance?

Now perform the model with an interaction term between PD and each variant. 

`NB`: E. coli clusters are calculated but not modeled because of the sample 
collection and preservation methods: room temperature without buffer induced
growth of *E. coli* (as well as other aerophilic and mesophilic bacteria, 
potentially). This reduces the number of independent tests to adjust for.

Also, one can argue that, since we did not see associations between variants and
clusterSums other than Bifidos and Fiber degraders, that the replication 
could be limited to 2 clusters. Same reasoning could go with variants that were
never associated with any cluster. For future modifications, we report the raw 
P-values in the associations table.

$$
Clustersum \sim Age + Seq.Depth + PD + VariantDosage_k + PD \times VariantDosage_k
$$

```{r bugs and genetics with lm}
# Here you can choose to restrict your analysis to specific clusters
clusters_to_test <- colnames(select(cluster_sums.df, -SubjectID, `E. coli-Klebsiella`))

lm_results_interaction.df <- data_alltogether.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
#  filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 5 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + Case_status + dosage + dosage*Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()


# number of tests to correct for as bonferroni: number of cluster + variants tested
n_tests <- length(clusters_to_test) * n_indep_variants

lm_results_interaction_bonferoni.df <- lm_results_interaction.df %>% 
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni > 1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_interaction_bonferoni.df, file = file.path(default_outdir, "Q1", "01_lm_interaction_Variant.x.PD.tsv"))

```

We correct p-values using the Bonferroni method accounting for 
`r length(clusters_to_test)` clusters tested against `r n_indep_variants` 
independent variants as defined above, for a total of `r n_tests` independent 
tests. Then, we consider associations between clusters and variants with 
`Adj. P-value < 0.05` as significant further annotating whether each association 
passed family-wise error rate adjustment.

```{r bugs and genetics results table}
lm_results_interaction_bonferoni.df %>%
  filter(grepl("dosage", term)) %>% 
  arrange(desc(abs(estimate))) %>% 
  select(-statistic) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(filterable = TRUE, searchable = TRUE, sortable = TRUE, columns = list(variant = reactable::colDef(width = 200), term = reactable::colDef(width = 200)))
```

### Barplots of beta values assoc. with dosage only (blue) or dosage x PD (gold)

```{r}

vgsub <- Vectorize(gsub)

lm_results_interaction_bonferoni.df <- mutate(lm_results_interaction_bonferoni.df, term = vgsub("dosage", variant, term)) %>% 
  select(-variant)

data_plot.tmp <- lm_results_interaction_bonferoni.df %>% 
  filter(grepl("chr", term)) %>%
  mutate(
    variant = str_extract(term, "chr.*"),
    PD = ifelse(grepl("PD", term), "PD", "Control"), 
    term = NULL) 

plotly::ggplotly(
ggplot() +
  geom_col(data = data_plot.tmp, mapping = aes(x = variant, y = estimate, fill = PD), position = position_dodge()) +
  geom_text(data = data_plot.tmp %>% filter(p.value < 0.05, p.adj_bonferroni >= 0.1), mapping = aes(x = variant, y = estimate, label = "*")) +
  geom_text(data = data_plot.tmp %>% filter(p.adj_bonferroni < 0.1), mapping = aes(x = variant, y = estimate, label = "**")) +
  scale_fill_manual(values = c("skyblue3", "gold3")) +
  facet_wrap(~ clusterName) +
  geom_hline(yintercept = 0, color = "gray30") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(
    y = "Beta estimates", 
    title = "Full Betas range"
  ) 
)
```

Above is a barplot of beta estimates of bacterial clusters in relation to genotype dosages in the case of PD (gold) and health controls (blue). Significant associations were plotted as follows: * for p-value < 0.05, **: Bonferroni-adjusted P-value < 0.1.

### Boxplots of ClusterSum abundance vs PD and variants (p-value < 0.05)

Only a few associations are significant here. Below you can find all significant
ones in boxplots taking into consideration genotypes and PD case/control status.
Plots are sorted by decreasing absolute $\beta$ values.

```{r, fig.height=12, fig.width=8}
plot_top_assoc <- lm_results_interaction_bonferoni.df %>%
  filter(grepl("chr[0-9]", term), p.value < 0.05) %>% 
  arrange(desc(abs(estimate)))

ggarrange(plotlist = lapply(1:nrow(plot_top_assoc), function(i) {
  clst <- plot_top_assoc$clusterName[i]
  varnt <- str_extract(plot_top_assoc$term[i], "chr.*")
  bonfSignif <- plot_top_assoc$p.adj_bonferroni[i] < 0.1
  
  majAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.) - 1]
  minAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.)]
  tmp <- data_alltogether.df %>% set_names(make.names(colnames(.)))
  
  tmp[,varnt] <- case_when(
        round(tmp[[varnt]], 0) == 0 ~ paste(majAll, majAll, sep = "|"),
        round(tmp[[varnt]], 0) == 1 ~ paste(majAll, minAll, sep = "|"),
        round(tmp[[varnt]], 0) == 2 ~ paste(minAll, minAll, sep = "|"),
        ) %>% 
    factor(levels = c(paste(majAll, majAll, sep = "|"), paste(majAll, minAll, sep = "|"), paste(minAll, minAll, sep = "|")))
  
  return(ggplot(tmp,
         aes(
           x = eval(
             parse(text = varnt)
           ),
           y = eval(
             parse(text = make.names(clst)
           )),
           color = Case_status
         )) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2)) +
    geom_boxplot(alpha = 0.4,
                 show.legend = FALSE,
                 outliers = FALSE) +
    scale_color_manual(values = c("skyblue3", "gold3")) +
    labs(
      x = "Genotype Dosages (rounded)",
      y = "ClusterSum (crude)",
      title = paste(clst, "vs", varnt),
      color = "Disease status",
      caption = paste("Bonferroni significant:", bonfSignif)
    )
  )
  }), ncol = 2, nrow = 4,
  common.legend = TRUE)
```

## Q2 - is SNP VAF associated with cluster abundance in PD?

Here the model is restricted to `PD` subjects, adjusting for the same
set of confounders. See model below:

$$
ClusterSum_{PD} \sim Age_{PD} + Seq.Depth_{PD} + VariantDosage_{PD}
$$

Additional condition: We must have at least 5 participants carrying each genotype.

```{r}
lm_results_in_PD.df <- data_alltogether.df %>%
  filter(Case_status == "PD") %>% 
  pivot_longer(
    cols = colnames(select(per_sample_dosages.df, -SubjectID)), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + dosage, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()

write_tsv(lm_results_in_PD.df, file = file.path(default_outdir, "Q2", "01_lm_dosage_in_PD.tsv"))
```

## Q3 - is SNP VAF associated with cluster abundance in Controls?

Here the model is restricted to `Control` subjects, adjusting for the same
set of confounders. See model below:

$$
ClusterSum_{Control} \sim Age_{Control} + Seq.Depth_{Control} + VariantDosage_{Control}
$$

Additional condition: We must have at least 5 participants carrying each genotype.

```{r}
lm_results_in_Control.df <- data_alltogether.df %>%
  filter(Case_status == "Control") %>% 
  pivot_longer(
    cols = colnames(select(per_sample_dosages.df, -SubjectID)), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + dosage, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()

write_tsv(lm_results_in_Control.df, file = file.path(default_outdir, "Q3", "01_lm_dosage_in_Controls.tsv"))
```


# Approach 2 - Full MWAS followed by Gene Set Enrichment Analysis on same clusters

**Hypothesis**: Summing bacterial abundances to generate clusters, followed by `lm`
may introduce both biological and modeling biases. A possible biological bias
could be that one bacterium is strongly affected by the interaction, but others 
are only mildly, which would weaken the association. A technical bias could be 
the non-normality and zero inflation of the outcome (bacterial clusters), which 
with ordinary least squares regression could inflate type I or II errors 
depending on the case. 

**Aim**: To test all bacteria individually for interaction with variant and PD, 
annotating them later to each cluster.

**Method**: Gene set enrichment analysis (GSEA) Takes the strength of assocation 
between all bacteria/genes and an exposure (eg: interaction `PD x variant`), 
ranks them from higher to lower, and tests whether it is more likely to find 
more or less of them belonging to a grouping. This grouping could be fully agnostic
(e.g.: pathways, taxa, metabolic functions) or could be manually annotated 
(Haydeh's bacterial clusters). The more bacteria in the same cluster are in 
extreme ranks (top or bottom), the more the bacterial cluster/gene set/pathway
is significantly enriched or depleted, respectively. In practice, my approach is:

  1. [linDA](https://doi.org/10.1186/s13059-022-02655-5) to test **ALL** 
  bacteria (prevalence >= 5%) for association with `variant x PD` interaction. 
  2. Extract `t-statistic` as metrics for strength of association ($\beta$ 
  estimates could work too, as well as $\beta \times -log_{10}Pvalue$)
  3. Use [fgsea](https://doi.org/10.1101/060012) with statistic in `2.` and 
  bacteria groupings (by Haydeh) to test for enrichment. fgsea is available in
  [Bioconductor](https://bioconductor.org/packages/release/bioc/html/fgsea.html).

## Q1 - Interaction PD x variant GSEA

Significant associations in the `dosage:PD` term mean that the interaction between
PD and genetic dosage is associated with bacterial abundance.

Building on these results, a significant GSEA results for either of the clusters
tested would mean that there is a number of bacteria enriched/depleted together 
belonging to the same cluster.

```{r}
samples_to_keep.df <- colData(NGRC_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame()

variants_to_keep_rowData.df <- as.data.frame(rowData(NGRC_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(NGRC_data.mae[["genotypes_dosages.se"]])$GENE),])

per_sample_dosages.df <- t(assay(NGRC_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

meta_with_variants <- inner_join(per_sample_dosages.df, samples_to_keep.df, by = "SubjectID")

```

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(meta_with_variants), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[,meta_with_variants$SubjectID],
         meta.dat = meta_with_variants,
  formula = paste("~ Age_at_collection + seqs_scaled + Case_status +", gt, "+", paste(gt, "Case_status", sep = "*")), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names(c("variantOnly", "interactionPD")) %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q1", "02_LinDA_regression_dosage.x.PD_FullResults.tsv"))
```

### GSEA on variant-PD interactions {.tabset}

```{r}
# fgsea on variant only effects
variant_PD_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 5) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats)) %>% 
  lapply(function(x){
    x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
    x <- rename(x, microbial_cluster = pathway)
    return(x)
  })

gsea_on_variant_PD_nice.df <- bind_rows(gsea_on_variant_PD, .id = "snp_variant") %>% 

write_tsv(gsea_on_variant_PD_nice.df, path = file.path(default_outdir, "Q1", "03_full_GSEA_results_interaction.tsv"))
```

```{r, results="asis"}
gsea_on_variant_PD_someSignif <- gsea_on_variant_PD[sapply(gsea_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(gsea_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_PD_t_stat[[i]],
      fgseaRes = gsea_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

### GSEA on variant-only effects {.tabset}

These results confirm our findings above, showing the tendency of enrichment of 
Fiber Degraders in case of dosage only effect, while for the same variant, there 
is an opposite trend. 

This is particularly visible in variant `rs983361`, named chr4_89840793

```{r}
# fgsea on variant only effects
variant_only_t_stats <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 4) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats)) %>% 
  lapply(function(x){
    x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
    x <- rename(x, microbial_cluster = pathway)
    return(x)
  })

gsea_on_variant_PD_nice.df <- bind_rows(gsea_on_variant_PD, .id = "snp_variant") %>% 

write_tsv(gsea_on_variant_PD_nice.df, path = file.path(default_outdir, "Q1", "03_full_GSEA_results_interaction.tsv"))

```

```{r, results="asis"}
gsea_on_variant_PD_someSignif <- gsea_on_variant_PD[sapply(gsea_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(gsea_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_only_t_stats[[i]],
      fgseaRes = gsea_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

## Q2 - Variant-bacteria association in PD

```{r}
samples_to_keep.df <- colData(NGRC_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() 

variants_to_keep_rowData.df <- as.data.frame(rowData(NGRC_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(NGRC_data.mae[["genotypes_dosages.se"]])$GENE),])

per_sample_dosages.df <- t(assay(NGRC_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

meta_with_variants <- inner_join(per_sample_dosages.df, samples_to_keep.df, by = "SubjectID") %>% 
  filter(Case_status == "PD")

```

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(meta_with_variants), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[,meta_with_variants$SubjectID],
         meta.dat = meta_with_variants,
  formula = paste("~ Age_at_collection + seqs_scaled + ", gt), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names("variantOnly") %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q2", "02_LinDA_regression_dosage_in_PD.tsv"))
```

### GSEA on variant effect {.tabset}

```{r}
# fgsea on variant only effects
variant_PD_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 3) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats)) %>% 
  lapply(function(x){
    x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
    x <- rename(x, microbial_cluster = pathway)
    return(x)
  })

gsea_on_variant_PD_nice.df <- bind_rows(gsea_on_variant_PD, .id = "snp_variant")

write_tsv(gsea_on_variant_PD_nice.df, path = file.path(default_outdir, "Q2", "03_full_GSEA_results_variant_in_PD.tsv"))
```

```{r, results="asis"}
gsea_on_variant_PD_someSignif <- gsea_on_variant_PD[sapply(gsea_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(gsea_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_PD_t_stat[[i]],
      fgseaRes = gsea_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

## Q3 - Variant-bacteria association in Controls

```{r}
samples_to_keep.df <- colData(NGRC_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() 

variants_to_keep_rowData.df <- as.data.frame(rowData(NGRC_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(NGRC_data.mae[["genotypes_dosages.se"]])$GENE),])

per_sample_dosages.df <- t(assay(NGRC_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

meta_with_variants <- inner_join(per_sample_dosages.df, samples_to_keep.df, by = "SubjectID") %>% 
  filter(Case_status == "Control")

```

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(meta_with_variants), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[,meta_with_variants$SubjectID],
         meta.dat = meta_with_variants,
  formula = paste("~ Age_at_collection + seqs_scaled + ", gt), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names("variantOnly") %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q3", "02_LinDA_regression_dosage_in_Controls.tsv"))
```

### GSEA on variant effect {.tabset}

```{r}
# fgsea on variant only effects
variant_Control_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 3) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variant_Control <- lapply(variant_Control_t_stat, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats)) %>% 
  lapply(function(x){
    x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
    x <- rename(x, microbial_cluster = pathway)
    return(x)
  })

gsea_on_variant_Control_nice.df <- bind_rows(gsea_on_variant_Control, .id = "snp_variant")

write_tsv(gsea_on_variant_Control_nice.df, path = file.path(default_outdir, "Q3", "03_full_GSEA_results_variant_in_Controls.tsv"))
```

```{r, results="asis"}
gsea_on_variant_Control_someSignif <- gsea_on_variant_Control[sapply(gsea_on_variant_Control, function(x) any(x$padj < 0.05))]

for (i in names(gsea_on_variant_Control_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_Control_t_stat[[i]],
      fgseaRes = gsea_on_variant_Control_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

# Old code, re-do for each question 
## LinDA on SNCA and HLA

The model done here is identical to the first approach:

$$
Bacterium_i \sim Age + Seq.Depth + PD + Variant_k + PD \times Variant_k
$$

And we extract coefficients (logFC) in relation to **dosage only** and 
**dosage in PD** to perform GSEA using Haydeh's clusters







# Approach 2.5 - GSEA on attributes found in `bugphyzz` (WIP)

This approach is even more unbiased. The analysis is based on their vignette.
(https://www.bioconductor.org/packages/release/data/experiment/vignettes/bugphyzz/inst/doc/bugphyzz.html)

```{r}
# BiocManager::install("bugphyzz") 
library(bugphyzz)
bp <- importBugphyzz()

names(bp)
```

There are many `data.frame`s with bug classification, with some overlap expected.
These sets can be used to test enrichment of a particular trait. In particular,
aerophilicity, gram staining, pathogenicity and butyrate production could be 
of interest.

Mapping can be done using `Taxon_name` and `Rank`

```{r}
bp$`butyrate-producing bacteria`

```

As can be seen, the list is not too large, in this case, and mapping is hard.