---
title: "PD, microbiome and host genetics on SNCA and HLA variants on NGRC"
author: "Giacomo Antonello, Zachary D. Wallen, Charles F. Murchinson, Levi Waldron, Haydeh Payami"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
---

```{r, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE
)
```

```{r}
default_outdir <- "output_NGRC/"

dir.create(default_outdir)
dir.create(file.path(default_outdir, "Q1"))
dir.create(file.path(default_outdir, "Q2"))
dir.create(file.path(default_outdir, "Q3"))

```

```{r, include=FALSE}

# Check and install necessary CRAN packages
CRAN_required_packages <- c("tidyverse", "gwasrapidd", "DT", "kableExtra", "data.table", "ggpubr", "BiocManager", "readxl","kableExtra", "plotly", "modeest")

install.packages(
  CRAN_required_packages[!(CRAN_required_packages %in% installed.packages())] 
)

# Check and install necessary Bioconductor packages
BioC_required_packages <- c("Maaslin2", "ANCOMBC", "biomaRt", "mia", "phyloseq", "microbiome","SIAMCAT", "fgsea")
BioC_required_packages <- BioC_required_packages[!(BioC_required_packages %in% installed.packages())]

if(length(BioC_required_packages) > 0) {
  BiocManager::install(BioC_required_packages[!(BioC_required_packages %in% installed.packages())],
                       ask = FALSE,
                       update = TRUE)
}


# load basic data manipulation packages
library(DT)
library(kableExtra)
library(MultiAssayExperiment)
library(ggpubr)
library(tidyverse)

theme_set(theme_light())
# load previously wrangled data
load("../Data/NGRC_data.mae.RData")
  
rm(BioC_required_packages,CRAN_required_packages)
gc(verbose = FALSE)
```

Below is the data for the NGRC dataset. Data have been 
further prepared by Giacomo to make one single data object that would hopefully, 
speed up the preparation phase. This data object contains microbiome, pathway, 
participants metadata and genetic data.

```{r}
NGRC_data.mae
```

Genetic data have been retrieved with the protocol available in `Data/genetic_data_preparation.{Rmd,html}`.
In brief:

  1. The GWAS catalog was searched for variants associated with the ontology term
  `MONDO_0005180`, and downloaded. all this with 
  [gwasrapidd](https://rmagno.eu/gwasrapidd/articles/gwasrapidd.html). 
  Catalog's version is not stated, I'm assuming that since the packages uses the `REST API`, 
  it should be the newest.
  
  2. Gene annotation was taken from `biomaRt`, but the GWAS catalog had annotations
  too. Naming is based on the HUGO nomenclature.

# Approach 1 - Sum bacteria in each cluster together

## Define `ClusterSum`

Haydeh has manually curated a list of bacteria that go together into 6 different
clusters elevated or decreased in PD. To see the genetic impact of PD associated
variants on these bacteria without having to pay a high multiple testing price.

Our first approach is to do the following:

  1. Define which species belong to which cluster
  2. Divide % abundances by 100, to obtain relative abundances again
  3. Sum relative abundances of all species within each cluster

```{r}
biomarkers_medrxiv.df <- readxl::read_xlsx("../Data/Biomarkers_Payami_medRxiv.xlsx")

datatable(biomarkers_medrxiv.df %>% select(-species_metaphlan, `Number of connections`), caption = "List of Bacterial species and their cluster assignment as manually curated by prof. Haydeh Payami")
```

```{r}

cluster_sums.df <- left_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(NGRC_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  split.data.frame(f = .$cluster_name) %>% 
  lapply(select, all_of(MultiAssayExperiment::colnames(NGRC_data.mae)[["metaphlan_RelAbund.se"]])) %>% 
  lapply(colSums) %>% 
  bind_cols(SubjectID = names(.[[1]])) %>% 
  relocate(SubjectID)

cluster_sums.df %>% 
  mutate_if(is.numeric, round, 3) %>% 
datatable(caption = "Snippet of values obtained with the sum method across clusters. These will be the variables we will use to test association with variants' dosages")

```

For each cluster, see how many species were correctly mapped to the NGRC dataset.

```{r}

n_all_species <- biomarkers_medrxiv.df %>% 
  count(cluster_name)

n_mapped_species <- inner_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(NGRC_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  count(cluster_name)

unmapped_species <- biomarkers_medrxiv.df %>%
  mutate(Species2 = case_when(!(species_metaphlan %in% rownames(
    assay(NGRC_data.mae, "metaphlan_RelAbund.se")
  )) ~ Species,
  TRUE ~ "")) %>% group_by(cluster_name) %>% reframe(unmapped = paste0(Species2[Species2 != ""], collapse = ", "))

full_join(n_all_species, n_mapped_species, by = "cluster_name") %>% 
  full_join(unmapped_species, by = "cluster_name") %>% 
  set_names(c("Cluster Name", "Total", "N. Mapped", "Not Mapped")) %>% 
  arrange(desc(Total)) %>% 
  kbl(caption = "Number of total species and correctly mapped species in the NGRC dataset")

```

These variables are far from normal, and log transformation does not help. Maybe
center-log ratio could help, but it can introduce biases on low-abundance bacteria
and clusters.

```{r, fig.cap= "Distribution of cumulative sum of relative abundances of cluster member species.", fig.height=6, fig.width=8}
cluster_sums_raw.mtx <-  cluster_sums.df %>%
  column_to_rownames("SubjectID") %>% 
  as.matrix() %>% 
  reshape2::melt(value.name = "raw")

cluster_sums_sqrt.mtx <-  cluster_sums.df %>%
  column_to_rownames("SubjectID") %>%
  apply(2, sqrt) %>% 
  reshape2::melt(value.name = "sqrt")

cluster_sums_log.mtx <-  cluster_sums.df %>%
  column_to_rownames("SubjectID") %>%
  apply(2, function(x) (log(x + 1))) %>% 
  reshape2::melt(value.name = "log_plus_1")

cluster_sums_clr.df <-  cluster_sums.df %>%
  column_to_rownames("SubjectID") %>%
  apply(2, compositions::clr) %>% 
  reshape2::melt(value.name = "clr")

data_transf_together <- inner_join(cluster_sums_raw.mtx, cluster_sums_sqrt.mtx, by = c("Var1", "Var2")) %>% 
  inner_join(cluster_sums_log.mtx, by = c("Var1", "Var2")) %>% 
  inner_join(cluster_sums_clr.df, by = c("Var1", "Var2")) %>% 
  rename(
    SubjectID = Var1,
    clusterName = Var2
  )

data_transf_together %>% 
  pivot_longer(cols = colnames(data_transf_together)[3:ncol(data_transf_together)], 
               names_to = "Transform", values_to = "transf_abund") %>% 
  mutate(Transform = factor(Transform, levels = colnames(data_transf_together)[3:ncol(data_transf_together)])) %>% 
  ggplot(aes(y = transf_abund, color = clusterName)) +
  geom_boxplot(fill = "transparent") +
  scale_color_manual(values = ggsci::pal_jco()(6)) +
  facet_wrap(~ Transform, scale = "free_y")

```

## Define variants and samples to test, and multiple testing cost

Genetic variants were included if retained if:

  - they were associated with PD in any GWAS catalog entry (P < 5e-8)
  - they were found in the TOPMed imputed genotype data
  - they had a gene name associated to them (for easy lookup, but exception may
  be missed)

```{r}

samples_to_keep.df <- colData(NGRC_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() %>% 
  filter(
    !is.na(Age_at_collection),
    !is.na(seqs_scaled),
    !is.na(Case_status)
    )

variants_to_keep_rowData.df <- as.data.frame(rowData(NGRC_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(NGRC_data.mae[["genotypes_dosages.se"]])$GENE),])
rownames(variants_to_keep_rowData.df) <- variants_to_keep_rowData.df$RSID

per_sample_dosages.df <- t(assay(NGRC_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

data_alltogether_Q1.df <- inner_join(
  # clustered bacteria variables
  cluster_sums.df %>% mutate_if(is.numeric, compositions::clr), 
  # variants available to test
  per_sample_dosages.df, by = "SubjectID") %>% 
  inner_join(
  # other individual data (metadata), excluding unwanted samples
    samples_to_keep.df, by = "SubjectID")

as_tibble(variants_to_keep_rowData.df) %>% 
  set_names(
    gsub("INFO_", "", colnames(.))
    ) %>%
  rename(
    `Empirical (R2)` = ER2,
    `Imputation Quality (R2)` = R2,
         ) %>% 
  datatable(caption = "List of variants found in the GWAS catalog that are explicitly mapped to the SNCA or HLA genes.")
```

```{r LDlink parameters setup}
r2_thresh <- 0.1
refPop <- "CEU"
maf_thresh <- 0.01
```

Some of these variants are in high LD. We want to adjust for the total number of 
independent tests, therefore we adjust for variants in low LD only across the 
genes of interest. We used an R API based on NIH's tool `LDlink`. Parameters
chosen are:

  - r2 threshold = `r r2_thresh`
  - reference population (1000 Genomes) = `r refPop`
  - MAF threshold = `r maf_thresh`

```{r}

cache(
  variants_pruning_step <- split.data.frame(variants_to_keep_rowData.df,variants_to_keep_rowData.df$CHROM) %>% 
  lapply( 
    function(x)
    LDlinkR::SNPclip(
      r2_threshold = r2_thresh,
      snps = x$RSID,
      pop = refPop,
      maf_threshold = maf_thresh,
      token = "78e50d5fd7e7", 
      genome_build = "grch38"
      )
  ), dir = file.path(default_outdir, "variant_pruning_cache")
  )

variants_pruning_table <- tryCatch(lapply(
  variants_pruning_step, mutate, Details = ifelse(grepl("removed", Details), "Removed", "Kept")
  ) %>% 
  lapply(group_by, Details) %>%
  lapply(tally) %>%
  bind_rows(.id = "Chromosome") %>% 
  pivot_wider(names_from = "Details", values_from = "n") %>% 
  mutate(Total = Kept + Removed),
  error = function(e) return(NA))

kbl(variants_pruning_table, caption = paste0("Number of idependent variants ('Kept') based on LDlink calculations on default parameters (R2 = ", r2_thresh, ", MAF = ", maf_thresh, "). The population chosen is 'CEU'. The sum of 'Kept' values represents the number of independent tests that need to be accounted for later on. NB: All variants will be tested, but the number of independent tests will be based on this 'Kept' column.")) %>% 
  kable_styling()
```

```{r define number of tests to account for}
n_indep_variants <- ifelse(all(is.na(variants_pruning_table)), 6, sum(sapply(variants_pruning_step, function(x) sum(grepl("kept", x$Details)))))

# Here you can choose to restrict your analysis to specific clusters
clusters_to_test <- colnames(select(cluster_sums.df, -SubjectID))
n_tests <- length(clusters_to_test) * n_indep_variants
  
cbind(
  c(
    "Intependent Variants",
    "Clusters tested",
    "Total independent test"
  ),
  c(n_indep_variants, length(clusters_to_test), n_tests)
) %>% 
  kbl(caption = "Overview of the number of independent variants and clusters tested that result in the total number of independent tests. The test is done multiplying all sources of independent tests.") %>% 
  kable_styling()
```

## Sample inclusion criteria and statistical modeling

  - Available data for age, sequencing depth, microbiome PD cumulative scores 
  (N unchanged)
  
  - Available genetic data (N goes from `r nrow(samples_to_keep.df)` to `r nrow(data_alltogether_Q1.df)`)

Below is some information of the variants associated with PD found in our 
imputed genotypes.

```{r}
cbind(data_alltogether_Q1.df %>% 
  select(starts_with("chr")) %>% 
  mutate_all(round, 0) %>% 
  sapply(table) %>% 
  t(), 
  data_alltogether_Q1.df %>% 
    select(starts_with("chr")) %>% 
    mutate_all(round, 0) %>% 
    sapply(function(x) table(x) %>% prop.table()) %>% 
    t()) %>% 
  .[,c(1,4,2,5,3,6)] %>%   
  as.data.frame() %>% 
  set_names(make.unique(colnames(.))) %>% 
  rownames_to_column("id") %>%
  separate_wider_delim(cols = id, names = c("CHROM", "POS", "REF", "ALT"), delim = "_", too_few = "align_start") %>% 
  rename(
    `0 (N)` = `0`,
    `0 (%)` = `0.1`,
    `1 (N)` = `1`,
    `1 (%)` = `1.1`,
    `2 (N)` = `2`,
    `2 (%)` = `2.1`
  ) %>% 
  mutate_at(.vars = vars(c(`0 (%)`,`1 (%)`,`2 (%)`)), .funs = function(x) round(x*100, 2)) %>% 
  
  kbl(caption = "Number of samples with alternate allele dosages. NB: no distinction is done for 1|0 or 0|1, as both carry one alternate allese. Values are shown as total counts and percentages") %>% 
  kable_styling() %>% 
  add_header_above(c("Variant Information" = 4, "Distribution of dosages in the cohort studied" = 6)) %>% 
  column_spec(c(4,6,8), border_right = TRUE)
```

## Q1 - Is there interaction between SNP variant allele dosage and PD status (case or control) on cluster abundance?

`NB`: E. coli clusters are calculated but not modeled because of the sample 
collection and preservation methods: room temperature without buffer induced
growth of *E. coli* (as well as other aerophilic and mesophilic bacteria, 
potentially). This reduces the number of independent tests to adjust for.

Also, one can argue that, since we did not see associations between variants and
clusterSums other than Bifidos and Fiber degraders in the discovery sataset, 
that the replication could be limited to 2 clusters. Same reasoning could go 
with variants that were never associated with any cluster. For future 
modifications, we report the raw P-values in the associations table.



$$
Clustersum \sim Age + Seq.Depth + PD + VariantDosage_k + PD \times VariantDosage_k
$$

```{r bugs and genetics with lm}

lm_results_interaction.df <- data_alltogether_Q1.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + stool_travel_time + Case_status + dosage + dosage*Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup() %>% 
  # Adjust p-values to account for the number of independent tests
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni > 1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_interaction.df, file = file.path(default_outdir, "Q1", "01_lm_interaction_Variant.x.PD.tsv"))

```

We correct p-values using the Bonferroni method accounting for 
`r length(clusters_to_test)` clusters tested against `r n_indep_variants` 
independent variants as defined above, for a total of `r n_tests` independent 
tests. Then, we consider associations between clusters and variants with 
`Adj. P-value < 0.05` as significant further annotating whether each association 
passed family-wise error rate adjustment.

```{r bugs and genetics results table}
lm_results_interaction.df %>%
  filter(grepl("dosage", term)) %>% 
  arrange(desc(abs(estimate))) %>% 
  select(-statistic) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(filterable = TRUE, searchable = TRUE, sortable = TRUE, columns = list(clusterName = reactable::colDef(width = 180), variant = reactable::colDef(width = 180), term = reactable::colDef(width = 180)))
```

### Barplots of beta values assoc. with dosage only (blue) or dosage x PD (gold)

```{r, fig.cap="Barplots of beta estimates of bacterial clusters in relation to genotype dosages in the case of PD (gold) and health controls (blue). Significant associations were plotted as follows: * for p-value < 0.05, **: Bonferroni-adjusted P-value < 0.1."}

vgsub <- Vectorize(gsub)

lm_results_interaction.df <- mutate(lm_results_interaction.df, term = vgsub("dosage", variant, term)) %>% 
  select(-variant)

data_plot.tmp <- lm_results_interaction.df %>% 
  filter(grepl("chr", term)) %>%
  mutate(
    variant = str_extract(term, "chr.*"),
    PD = ifelse(grepl("PD", term), "PD", "Control"), 
    term = NULL) 

plotly::ggplotly(
ggplot() +
  geom_col(data = data_plot.tmp, mapping = aes(x = variant, y = estimate, fill = PD), position = position_dodge()) +
  geom_text(data = data_plot.tmp %>% filter(p.value < 0.05, p.adj_bonferroni >= 0.1), mapping = aes(x = variant, y = estimate, label = "*")) +
  geom_text(data = data_plot.tmp %>% filter(p.adj_bonferroni < 0.1), mapping = aes(x = variant, y = estimate, label = "**")) +
  scale_fill_manual(values = c("skyblue3", "gold3")) +
  facet_wrap(~ clusterName) +
  geom_hline(yintercept = 0, color = "gray30") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(
    y = "Beta estimates", 
    title = "Full Betas range"
  ) 
)
```

### Boxplots of ClusterSum abundance vs PD and variants (p-value < 0.05)

Boxplots of associations with raw p-value , 0.05 (if any). If not, this step
will be skipped.

```{r}
plot_top_assoc <- lm_results_interaction.df %>%
  filter(grepl("chr[0-9]", term), p.value < 0.05) %>% 
  arrange(desc(abs(estimate)))
```

```{r, eval = nrow(plot_top_assoc) > 0, fig.height=12, fig.width=8}

ggarrange(plotlist = lapply(1:nrow(plot_top_assoc), function(i) {
  clst <- plot_top_assoc$clusterName[i]
  varnt <- str_extract(plot_top_assoc$term[i], "chr.*")
  bonfSignif <- plot_top_assoc$p.adj_bonferroni[i] < 0.1
  
  majAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.) - 1]
  minAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.)]
  tmp <- data_alltogether_Q1.df %>% set_names(make.names(colnames(.)))
  
  tmp[,varnt] <- case_when(
        round(tmp[[varnt]], 0) == 0 ~ paste(majAll, majAll, sep = "|"),
        round(tmp[[varnt]], 0) == 1 ~ paste(majAll, minAll, sep = "|"),
        round(tmp[[varnt]], 0) == 2 ~ paste(minAll, minAll, sep = "|"),
        ) %>% 
    factor(levels = c(paste(majAll, majAll, sep = "|"), paste(majAll, minAll, sep = "|"), paste(minAll, minAll, sep = "|")))
  
  return(ggplot(tmp,
         aes(
           x = eval(
             parse(text = varnt)
           ),
           y = eval(
             parse(text = make.names(clst)
           )),
           color = Case_status
         )) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2)) +
    geom_boxplot(alpha = 0.4,
                 show.legend = FALSE,
                 outliers = FALSE) +
    scale_color_manual(values = c("skyblue3", "gold3")) +
    labs(
      x = "Genotype Dosages (rounded)",
      y = "ClusterSum (crude)",
      title = paste(clst, "vs", varnt),
      color = "Disease status",
      caption = paste("Bonferroni significant:", bonfSignif)
    )
  )
  }), ncol = 2, nrow = 4,
  common.legend = TRUE)
```

## Q2 - is SNP VAF associated with cluster abundance in PD?

Here the model is restricted to `PD` subjects, adjusting for the same
set of confounders. See model below:

$$
ClusterSum_{PD} \sim Age_{PD} + Seq.Depth_{PD} + VariantDosage_{PD}
$$

```{r}
data_alltogether_Q2.df <- filter(data_alltogether_Q1.df, Case_status == "PD")
```

```{r}
lm_results_in_PD.df <- data_alltogether_Q2.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + stool_travel_time + dosage, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup() %>% 
  # Adjust p-values to account for the number of independent tests
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni > 1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_in_PD.df, file = file.path(default_outdir, "Q2", "01_lm_dosage_in_PD.tsv"))
```

**Table below**: List of cluster-variant dosage associations in PD . 
Numeric values have been rounded to the 4th digit. The 
`significant` column indicates whether the adjusted p-value is lower than 0.1.

```{r}
lm_results_in_PD.df %>% 
  filter(grepl("dosage", term)) %>% 
  select(-term) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(
    filterable = TRUE,
    searchable = TRUE,
    sortable = TRUE,
    columns = list(
      clusterName = reactable::colDef(width = 180),
      variant = reactable::colDef(width = 150)
    )
  )
```

## Q3 - is SNP VAF associated with cluster abundance in Controls?

Here the model is restricted to `Control` subjects, adjusting for the same
set of confounders. See model below:

$$
ClusterSum_{Control} \sim Age_{Control} + Seq.Depth_{Control} + VariantDosage_{Control}
$$

```{r}
data_alltogether_Q3.df <- filter(data_alltogether_Q1.df, Case_status == "Control")
```

```{r}
lm_results_in_Control.df <- data_alltogether_Q3.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = clusters_to_test, 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + stool_travel_time + dosage, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup() %>% 
  # Adjust p-values to account for the number of independent tests
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni > 1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_in_Control.df, file = file.path(default_outdir, "Q3", "01_lm_dosage_in_Controls.tsv"))
```

**Table below**: List of cluster-variant dosage associations in Controls. 
Numeric values have been rounded to the 4th digit. The `significant` column 
indicates whether the adjusted p-value is lower than 0.1.

```{r}
lm_results_in_Control.df %>% 
  filter(grepl("dosage", term)) %>% 
  select(-term) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(
    filterable = TRUE,
    searchable = TRUE,
    sortable = TRUE,
    columns = list(
      clusterName = reactable::colDef(width = 180),
      variant = reactable::colDef(width = 150)
    )
  )
```

# Approach 2 - Full MWAS followed by Microbe Set Enrichment Analysis on same clusters

**Hypothesis**: Summing bacterial abundances to generate clusters, followed by `lm`
may introduce both biological and modeling biases. A possible biological bias
could be that one bacterium is strongly affected by the interaction, but others 
are only mildly, which would weaken the association. A technical bias could be 
the non-normality and zero inflation of the outcome (bacterial clusters), which 
with ordinary least squares regression could inflate type I or II errors 
depending on the case. 

**Aim**: To test all bacteria individually for interaction with variant and PD, 
annotating them later to each cluster.

**Method**: Microbe Set Enrichment Analysis (MSEA) Takes the strength of assocation 
between all bacteria/genes and an exposure (eg: interaction `PD x variant`), 
ranks them from higher to lower, and tests whether it is more likely to find 
more or less of them belonging to a grouping. This grouping could be fully agnostic
(e.g.: pathways, taxa, metabolic functions) or could be manually annotated 
(Haydeh's bacterial clusters). The more bacteria in the same cluster are in 
extreme ranks (top or bottom), the more the bacterial cluster/gene set/pathway
is significantly enriched or depleted, respectively. In practice, my approach is:

  1. [linDA](https://doi.org/10.1186/s13059-022-02655-5) to test **ALL** 
  bacteria (prevalence >= 5%) for association with `variant x PD` interaction.
  For this method, we use metaphlan estimated counts.
  2. Extract `t-statistic` as metrics for strength of association ($\beta$ 
  estimates could work too, as well as $\beta \times -log_{10}Pvalue$)
  3. Use [fgsea](https://doi.org/10.1101/060012) with statistic in `2.` and 
  bacteria groupings (by Haydeh) to test for enrichment. fgsea is available in
  [Bioconductor](https://bioconductor.org/packages/release/bioc/html/fgsea.html).

**NB: Only variants with at least 1 significantly enriched cluster are shown.**
**Significance is defined as FDR Q-value < 0.05**

```{r}
prevalence_curve <- assay(NGRC_data.mae, "metaphlan_Counts.se") %>% 
  microbiome::prevalence() %>% 
  sort(decreasing = TRUE) 
```

Also, filtering by prevalence results in considerable loss of taxa: 
`r sum(prevalence_curve >= 0.05)` are retained, `r sum(prevalence_curve < 0.05)` 
are discarded. Note that **metaphlan_Counts** have 1 extra taxon compared to 
`metaphlan_RelAbund`, that is the `UNKNOWN` taxa. This will be modeled too and 
used as control, as we generally expect UNKNOWN taxa to encompass too many taxa 
to return a clean signal.

## Q1 - Interaction PD x variant MSEA

Significant associations in the `dosage:PD` term mean that the interaction between
PD and genetic dosage is associated with bacterial abundance.

Building on these results, a significant MSEA results for either of the clusters
tested would mean that there is a number of bacteria enriched/depleted together 
belonging to the same cluster.

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(data_alltogether_Q1.df), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[, data_alltogether_Q1.df$SubjectID],
         meta.dat = data_alltogether_Q1.df,
  formula = paste("~ Age_at_collection + seqs_scaled + Case_status +", gt, "+", paste(gt, "Case_status", sep = "*")), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names(c("variantOnly", "interactionPD")) %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q1", "02_LinDA_regression_dosage.x.PD_FullResults.tsv"))
```

### MSEA plot on interaction effects from interaction model {.tabset}

```{r}
# fgsea on variant only effects
variant_PD_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  #this index must be checked in case the formula in the modeling chunk above gets changed
  lapply("[[", 5) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

MSEA_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) {
  set.seed(1234)
  x <- fgsea(
    pathways = split(
      biomarkers_medrxiv.df$species_metaphlan,
      biomarkers_medrxiv.df$cluster_name
    ),
    stats = stats
  )
  x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
  return(x)
})

MSEA_on_variant_PD_nice.df <- bind_rows(MSEA_on_variant_PD, .id = "snp_variant") %>% 
  rename(clusterName = pathway)

write_tsv(MSEA_on_variant_PD_nice.df, file = file.path(default_outdir, "Q1", "03_full_MSEA_results_interaction.tsv"))
```

```{r, results="asis"}
MSEA_on_variant_PD_someSignif <- MSEA_on_variant_PD[sapply(MSEA_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(MSEA_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
        pathways = split(
          biomarkers_medrxiv.df$species_metaphlan,
          biomarkers_medrxiv.df$cluster_name
        ),
        stats = variant_PD_t_stat[[i]],
        fgseaRes = MSEA_on_variant_PD_someSignif[[i]]
      )
    )
    cat("\n")
}
```

### {-}

### MSEA plot on variant-only effects from interaction model {.tabset}

```{r}
# fgsea on variant only effects
variant_only_t_stats <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 4) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

MSEA_on_variant_PD <- lapply(variant_only_t_stats, function(stats) {
  set.seed(1234)
  x <- fgsea(
    pathways = split(
      biomarkers_medrxiv.df$species_metaphlan,
      biomarkers_medrxiv.df$cluster_name
    ),
    stats = stats
  )
  x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
  return(x)
})

MSEA_on_variant_PD_nice.df <- bind_rows(MSEA_on_variant_PD, .id = "snp_variant") %>% 
  rename(microbial_cluster = pathway)

write_tsv(MSEA_on_variant_PD_nice.df, path = file.path(default_outdir, "Q1", "03_full_MSEA_results_variantOnly.tsv"))

```

```{r, results="asis"}
MSEA_on_variant_PD_someSignif <- MSEA_on_variant_PD[sapply(MSEA_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(MSEA_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_only_t_stats[[i]],
      fgseaRes = MSEA_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

## Q2 - Variant dosage vs bacteria association in PD

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(data_alltogether_Q2.df), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[,data_alltogether_Q2.df$SubjectID],
         meta.dat = data_alltogether_Q2.df,
  formula = paste("~ Age_at_collection + seqs_scaled + ", gt), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names("variantOnly") %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q2", "02_LinDA_regression_dosage_in_PD.tsv"))
```

### MSEA plot on variant effect in PD {.tabset}

```{r}
# fgsea on variant only effects
variant_PD_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 3) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

MSEA_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) {
  set.seed(1234)
  x <- fgsea(
    pathways = split(
      biomarkers_medrxiv.df$species_metaphlan,
      biomarkers_medrxiv.df$cluster_name
    ),
    stats = stats
  )
  x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
  return(x)
})

MSEA_on_variant_PD_nice.df <- bind_rows(MSEA_on_variant_PD, .id = "snp_variant") %>% 
  rename(microbial_cluster = pathway)

write_tsv(MSEA_on_variant_PD_nice.df, path = file.path(default_outdir, "Q2", "03_full_MSEA_results_variant_in_PD.tsv"))
```

```{r, results="asis"}
MSEA_on_variant_PD_someSignif <- MSEA_on_variant_PD[sapply(MSEA_on_variant_PD, function(x) any(x$padj < 0.05))]

for (i in names(MSEA_on_variant_PD_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_PD_t_stat[[i]],
      fgseaRes = MSEA_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

## Q3 - Variant-bacteria association in Controls

### Regression 

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(data_alltogether_Q3.df), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(NGRC_data.mae, "metaphlan_Counts.se")[,data_alltogether_Q3.df$SubjectID],
         meta.dat = data_alltogether_Q3.df,
  formula = paste("~ Age_at_collection + seqs_scaled + ", gt), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names("variantOnly") %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "Q3", "02_LinDA_regression_dosage_in_Controls.tsv"))
```

### MSEA plot on variant effect in Controls {.tabset}

```{r}
# fgsea on variant only effects
variant_Control_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 3) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

MSEA_on_variant_Control <- lapply(variant_Control_t_stat, function(stats) {
  set.seed(1234)
  x <- fgsea(
    pathways = split(
      biomarkers_medrxiv.df$species_metaphlan,
      biomarkers_medrxiv.df$cluster_name
    ),
    stats = stats
  )
  x$leadingEdge <- sapply(x$leadingEdge, paste, collapse = "; ")
  return(x)
})

MSEA_on_variant_Control_nice.df <- bind_rows(MSEA_on_variant_Control, .id = "snp_variant") %>% 
  rename(microbial_cluster = pathway)

write_tsv(MSEA_on_variant_Control_nice.df, path = file.path(default_outdir, "Q3", "03_full_MSEA_results_variant_in_Controls.tsv"))
```

```{r, results="asis"}
MSEA_on_variant_Control_someSignif <- MSEA_on_variant_Control[sapply(MSEA_on_variant_Control, function(x) any(x$padj < 0.05))]

for (i in names(MSEA_on_variant_Control_someSignif)) {
    cat("\n####", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_Control_t_stat[[i]],
      fgseaRes = MSEA_on_variant_Control_someSignif[[i]])
      )
    cat("\n")
}
```

### {-}

# Results highlights

To be done

# Future directions

## Find other bacteria clusters from other sources (e.g. `bugphyzz`)

```{r}
# BiocManager::install("bugphyzz") 
library(bugphyzz)
bp <- importBugphyzz()

tibble(
  "Metabolic trait" = names(bp),
  "Number of bacteria in BugPhyzz" = sapply(bp, nrow)
           ) %>% 
  arrange(desc(`Number of bacteria in BugPhyzz`)) %>% 
  reactable::reactable(searchable = TRUE, filterable = TRUE)
```


## Increase set of variants to be tested

### FoxDen

```{r}
snp_data <- data.frame(
  rsID = c("rs11610045", "rs4698412", "rs823118", "rs11658976", "rs356182", 
             "rs2280104", "rs12456492", "rs11158026", "rs9261484", "rs10513789", 
             "rs76904798", "rs10797576", "rs2230288", "rs34637584", "rs80356773", 
             "rs34424986", "rs76763715"),
  Chromosome = c(12, 4, 1, 17, 4, 8, 18, 14, 6, 3, 12, 1, 1, 12, 1, 6, 1),
  Position = c(133063768, 15735148, 204929729, 46926463, 89724099, 
                      22642927, 43093415, 54851089, 30114775, 182762437, 
                      40614434, 232664611, 155235252, 40734202, 155235843, 
                      161768452, 155236246),
  Gene = c(
    "FBRSL1",
    "BST1",
    "NUCKS1",
    "WNT3",
    "SNCA",
    "BIN3",
    "RIT2",
    "GCH1",
    "TRIM40",
    "MCCC1",
    "LRRK2",
    "SIPA1L2",
    "GBA",
    "LRRK2",
    "GBA",
    "PARK2",
    "GBA"
  )
) %>% 
  mutate(
    AltName = ifelse(rsID == "rs76763715", "i4000415", ""),
    AlreadyInList = ifelse((rsID %in% rownames(rowData(NGRC_data.mae[["genotypes_dosages.se"]]))) | (AltName %in% rownames(rowData(NGRC_data.mae[["genotypes_dosages.se"]]))), "yes", "no")
  ) %>% 
  arrange(AlreadyInList)

kbl(snp_data, caption = "List of variants found in FoxDen data repository. Grayed rows indicate variants already in our selection criteria in the GWAS catalog. Variants grayed") %>% 
  kable_styling() %>% 
  row_spec(row = grep("yes", snp_data$AlreadyInList), color = "gray")
```

### Big table in Nalls et al. 2024, Lancet Neurology

Reference is [here](https://doi.org/10.1016/S1474-4422(19)30320-5)

```{r}
dataTable <- readxl::read_xlsx("../Data/GWAS_catalog_PD_variants/Nalls_2023_Table1Variants.xlsx") %>% 
  mutate(AlreadyInList = ifelse(rsID %in% rownames(rowData(NGRC_data.mae[["genotypes_dosages.se"]])), "yes", "no")) %>% 
  arrange(AlreadyInList) 

kbl(dataTable, caption = "List of variants associated with PD reported as 'novel' in a mega GWAS by the authors in Nalls E.A. et al, 2024 (https://doi.org/10.1016/S1474-4422(19)30320-5). Grayed rows indicate variants already in our selection criteria in the GWAS catalog. Variants grayed. Input table was semi-automatically extracted fromTable 1 of the publication, mistakes cannot be excluded.") %>% 
  kable_styling() %>% 
  row_spec(row = grep("yes", dataTable$AlreadyInList), color = "gray")
```
