---
title: "PD, microbiome and host genetics on SNCA and HLA variants"
subtitle: "Data: DoD"
author: "Giacomo Antonello, Zachary D. Wallen, Charles M. Murchinson, Levi Waldron, Haydeh Payami"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
---

```{r, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE
)

default_outdir <- "output_DoD/"
dir.create(default_outdir, showWarnings = FALSE)
```

```{r, include=FALSE}

# Check and install necessary CRAN packages
CRAN_required_packages <- c("tidyverse", "gwasrapidd", "DT", "kableExtra", "data.table", "ggpubr", "BiocManager", "readxl","kableExtra", "plotly", "modeest")

install.packages(
  CRAN_required_packages[!(CRAN_required_packages %in% installed.packages())] 
)

# Check and install necessary Bioconductor packages
BioC_required_packages <- c("Maaslin2", "ANCOMBC", "biomaRt", "mia", "phyloseq", "microbiome","SIAMCAT", "fgsea")
BioC_required_packages <- BioC_required_packages[!(BioC_required_packages %in% installed.packages())]

if(length(BioC_required_packages) > 0) {
  BiocManager::install(BioC_required_packages[!(BioC_required_packages %in% installed.packages())],
                       ask = FALSE,
                       update = TRUE)
}


# load basic data manipulation packages
library(DT)
library(kableExtra)
library(MultiAssayExperiment)
library(ggpubr)
library(tidyverse)

theme_set(theme_light())
# load previously wrangled data
load("../Data/DoD_data.mae.RData")

rm(BioC_required_packages,CRAN_required_packages)
gc(verbose = FALSE)
```

Below is the data sources for Wallen et al. 2022 (Nat Comms). Data have been 
further prepared by Giacomo to make one single data object that would hopefully, 
speed up the preparation phase. This data object contains microbiome, pathway, 
participants data and genetic data.

```{r}
DoD_data.mae
```

Genetic data have been retrieved with the protocol available in `Data/genetic_data_preparation.{Rmd,html}`.
In brief:

  1. The GWAS catalog was searched for variants associated with the ontology term
  `MONDO_0005180`, and downloaded. all this with 
  [gwasrapidd](https://rmagno.eu/gwasrapidd/articles/gwasrapidd.html). 
  Catalog's version is not stated, I'm assuming that since the packages uses the `REST API`, 
  it should be the newest.
  
  2. Gene annotation was taken from `biomaRt`, but the GWAS catalog had annotations
  too. Naming is based on the HUGO nomenclature.

# Assocation between PD markers and SNCA and HLA gene variants  (ClustSum)

## Define `ClusterSum`

Haydeh has manually curated a list of bacteria that go together into 6 different
clusters elevated or decreased in PD. To see the genetic impact of PD associated
variants on these bacteria without having to pay a high multiple testing price.

Our first approach is to do the following:

  1. Define which species belong to which cluster
  2. Divide % abundances by 100, to obtain relative abundances again
  3. Sum relative abundances of all species within each cluster

```{r}
biomarkers_medrxiv.df <- readxl::read_xlsx("../Data/Biomarkers_Payami_medRxiv.xlsx")

datatable(biomarkers_medrxiv.df %>% select(-species_metaphlan, `Number of connections`), caption = "List of Bacterial species and their cluster assignment as manually curated by prof. Haydeh Payami")
```

```{r}

cluster_sums.df <- left_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(DoD_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  split.data.frame(f = .$cluster_name) %>% 
  lapply(select, all_of(MultiAssayExperiment::colnames(DoD_data.mae)[["metaphlan_RelAbund.se"]])) %>% 
  lapply(colSums) %>% 
  bind_cols(sample_name = names(.[[1]])) %>% 
  relocate(sample_name)

cluster_sums.df %>% 
  mutate_if(is.numeric, round, 3) %>% 
datatable(caption = "Snippet of values obtained with the sum method across clusters. These will be the variables we will use to test association with variants' dosages")
```

For each cluster, see how many species were correctly mapped to the DoD dataset.

```{r}

n_all_species <- biomarkers_medrxiv.df %>% 
  count(cluster_name)

n_mapped_species <- inner_join(
  biomarkers_medrxiv.df,
  as.data.frame(assay(DoD_data.mae, "metaphlan_RelAbund.se")/100) %>%
    rownames_to_column("species_metaphlan"),
  by = "species_metaphlan"
  ) %>% 
  count(cluster_name)

unmapped_species <- biomarkers_medrxiv.df %>%
  mutate(Species2 = case_when(!(species_metaphlan %in% rownames(
    assay(DoD_data.mae, "metaphlan_RelAbund.se")
  )) ~ Species,
  TRUE ~ "")) %>% group_by(cluster_name) %>% reframe(unmapped = paste0(Species2[Species2 != ""], collapse = ", "))

full_join(n_all_species, n_mapped_species, by = "cluster_name") %>% 
  full_join(unmapped_species, by = "cluster_name") %>% 
  set_names(c("Cluster Name", "Total", "N. Mapped", "Not Mapped")) %>% 
  arrange(desc(Total)) %>% 
  kbl(caption = "Number of total species and correctly mapped species in the DoD dataset")

```

These variables are far from normal, and log transformation does not help.

```{r, fig.cap= "Distribution of cumulative sum of relative abundances of cluster member species."}
cluster_sums.df %>% 
  select(-sample_name) %>% 
  reshape2::melt() %>% 
  mutate(
    log2Transf = log2(value + 1),
    sq.root = sqrt(value)
    ) %>% 
  reshape2::melt(
    variable.name = "Transformation"
    ) %>% 
  mutate(
    Transformation = case_when(
      Transformation == "value" ~ 'untransformed',
      Transformation == "log2Transf" ~ 'log2',
      Transformation == "sq.root" ~ 'sq.root',
      TRUE ~ 'error'
      )
    ) %>% 
  ggplot(aes(x = value, color = Transformation)) + 
  geom_histogram(fill = "transparent") + 
  facet_wrap(~variable) 
```

## Define variants of interest and number of independent tests

Genetic variants were included if retained if:

  - they were associated with PD in any GWAS catalog entry (P < 5e-8)
  - they were found in the TOPMed imputed genotype data
  - they had a gene name associated to them (for easy lookup, but exception may
  be missed)

```{r}

samples_to_keep.df <- colData(DoD_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() %>% 
  filter(
    !is.na(Age_at_collection),
    !is.na(seqs_scaled),
    !is.na(Case_status),
    collection_method != "swab"
    )

variants_to_keep_rowData.df <- as.data.frame(rowData(DoD_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(DoD_data.mae[["genotypes_dosages.se"]])$GENE),])
rownames(variants_to_keep_rowData.df) <- variants_to_keep_rowData.df$RSID

per_sample_dosages.df <- t(assay(DoD_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("sample_name")

data_alltogether.df <- inner_join(
  # clustered bacteria variables
  cluster_sums.df, 
  # variants available to test
  per_sample_dosages.df, by = "sample_name") %>% 
  inner_join(
  # other individual data (metadata), excluding unwanted samples
    samples_to_keep.df, by = "sample_name")

as_tibble(variants_to_keep_rowData.df) %>% 
  set_names(
    gsub("INFO_", "", colnames(.))
    ) %>%
  rename(
    `Empirical (R2)` = ER2,
    `Imputation Quality (R2)` = R2,
         ) %>% 
  datatable(caption = "List of variants found in the GWAS catalog that are explicitly mapped to the SNCA or HLA genes.")
```

```{r LDlink parameters setup}
r2_thresh <- 0.1
refPop <- "CEU"
maf_thresh <- 0.01
```

Some of these variants are in high LD. We want to adjust for the total number of 
independent tests, therefore we adjust for variants in low LD only across the 
genes of interest. We used an R API based on NIH's tool `LDlink`. Parameters
chosen are:

  - r2 threshold = `r r2_thresh`
  - reference population (1000 Genomes) = `r refPop`
  - MAF threshold = `r maf_thresh`

```{r}

cache(
  variants_pruning_step <- split.data.frame(variants_to_keep_rowData.df,variants_to_keep_rowData.df$CHROM) %>% 
  lapply( 
    function(x)
    LDlinkR::SNPclip(
      r2_threshold = r2_thresh,
      snps = x$RSID,
      pop = refPop,
      maf_threshold = maf_thresh,
      token = "78e50d5fd7e7", 
      genome_build = "grch38"
      )
  ), dir = file.path(default_outdir, "variant_pruning_cache")
  )

variants_pruning_table <- tryCatch(lapply(
  variants_pruning_step, mutate, Details = ifelse(grepl("removed", Details), "Removed", "Kept")
  ) %>% 
  lapply(group_by, Details) %>%
  lapply(tally) %>%
  bind_rows(.id = "Chromosome") %>% 
  pivot_wider(names_from = "Details", values_from = "n") %>% 
  mutate(Total = Kept + Removed),
  error = function(e) return(NA))

kbl(variants_pruning_table, caption = paste0("Number of idependent variants ('Kept') based on LDlink calculations on default parameters (R2 = ", r2_thresh, ", MAF = ", maf_thresh, "). The population chosen is 'CEU'. The sum of 'Kept' values represents the number of independent tests that need to be accounted for later on. NB: All variants will be tested, but the number of independent tests will be based on this 'Kept' column.")) %>% 
  kable_styling()
```

```{r define number of tests to account for}
n_indep_variants <- ifelse(all(is.na(variants_pruning_table)), 6, sum(sapply(variants_pruning_step, function(x) sum(grepl("kept", x$Details)))))
```

## Sample inclusion criteria and statistical modeling

  - Available data for age, sequencing depth, microbiome PD cumulative scores (N unchanged)
  
  - Collection method should not be `swab` (N goes from  `r nrow(cluster_sums.df)` to `r nrow(samples_to_keep.df)`)
  
  - Available genetic data (N goes from `r nrow(samples_to_keep.df)` to `r nrow(data_alltogether.df)`)

Below is some information of the variants associated with PD found in our 
imputed genotypes.

```{r}
cbind(data_alltogether.df %>% 
  select(starts_with("chr")) %>% 
  mutate_all(round, 0) %>% 
  sapply(table) %>% 
  t(), 
  data_alltogether.df %>% 
    select(starts_with("chr")) %>% 
    mutate_all(round, 0) %>% 
    sapply(function(x) table(x) %>% prop.table()) %>% 
    t()) %>% 
  .[,c(1,4,2,5,3,6)] %>%   
  as.data.frame() %>% 
  set_names(make.unique(colnames(.))) %>% 
  rownames_to_column("id") %>%
  separate_wider_delim(cols = id, names = c("CHROM", "POS", "REF", "ALT"), delim = "_", too_few = "align_start") %>% 
  rename(
    `0 (N)` = `0`,
    `0 (%)` = `0.1`,
    `1 (N)` = `1`,
    `1 (%)` = `1.1`,
    `2 (N)` = `2`,
    `2 (%)` = `2.1`
  ) %>% 
  mutate_at(.vars = vars(c(`0 (%)`,`1 (%)`,`2 (%)`)), .funs = function(x) round(x*100, 2)) %>% 
  
  kbl(caption = "Number of samples with alternate allele dosages. NB: no distinction is done for 1|0 or 0|1, as both carry one alternate allese. Values are shown as total counts and percentages") %>% 
  kable_styling() %>% 
  add_header_above(c("Variant Information" = 4, "Distribution of dosages in the cohort studied" = 6)) %>% 
  column_spec(c(4,6,8), border_right = TRUE)
```

Now perform the model with an interaction term between PD and each variant.

$$
ClusterSum_i \sim Age + Seq.Depth + PD + Variant_k + PD \times Variant_k
$$

```{r bugs and genetics with lm}
lm_results_interaction.df <- data_alltogether.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = colnames(select(cluster_sums.df, -sample_name)), 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
#  filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 5 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + Case_status + dosage + dosage*Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()


# number of tests to correct for as bonferroni: number of cluster + variants tested
n_tests <- length(unique(lm_results_interaction.df$clusterName)) * n_indep_variants

lm_results_interaction_bonferoni.df <- lm_results_interaction.df %>% 
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni > 1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_interaction_bonferoni.df, file = file.path(default_outdir, "01_lm_interaction_Variant.x.PD.tsv"))

```

We correct p-values using the Bonferroni method accounting for `r 6` clusters tested
against `r n_indep_variants` independent variants as defined above, for a total of `r n_tests` 
independent tests. Then, we consider associations between clusters and variants 
with `Adj. P-value < 0.05` as significant further annotating whether each association 
passed family-wise error rate adjustment.

```{r bugs and genetics results table}
lm_results_interaction_bonferoni.df %>%
  filter(grepl("dosage", term)) %>% 
  arrange(desc(abs(estimate))) %>% 
  select(-statistic) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(filterable = TRUE, searchable = TRUE, sortable = TRUE, columns = list(variant = reactable::colDef(width = 200), term = reactable::colDef(width = 200)))
```

## Visualization of variant-cluster associations with P-value < 0.05

### Barplots of beta values assoc. with dosage only (blue) or dosage x PD (gold)

```{r}

vgsub <- Vectorize(gsub)

lm_results_interaction_bonferoni.df <- mutate(lm_results_interaction_bonferoni.df, term = vgsub("dosage", variant, term)) %>% 
  select(-variant)

data_plot.tmp <- lm_results_interaction_bonferoni.df %>% 
  filter(grepl("chr", term)) %>%
  mutate(
    variant = str_extract(term, "chr.*"),
    PD = ifelse(grepl("PD", term), "PD", "Control"), 
    term = NULL) 

plotly::ggplotly(
ggplot() +
  geom_col(data = data_plot.tmp, mapping = aes(x = variant, y = estimate, fill = PD), position = position_dodge()) +
  geom_text(data = data_plot.tmp %>% filter(p.value < 0.05, p.adj_bonferroni >= 0.1), mapping = aes(x = variant, y = estimate, label = "*")) +
  geom_text(data = data_plot.tmp %>% filter(p.adj_bonferroni < 0.1), mapping = aes(x = variant, y = estimate, label = "**")) +
  scale_fill_manual(values = c("skyblue3", "gold3")) +
  facet_wrap(~ clusterName) +
  geom_hline(yintercept = 0, color = "gray30") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(
    y = "Beta estimates", 
    title = "Full Betas range"
  ) 
)
```

Above is a barplot of beta estimates of bacterial clusters in relation to genotype dosages in the case of PD (gold) and health controls (blue). Significant associations were plotted as follows: * for p-value < 0.05, **: Bonferroni-adjusted P-value < 0.1.

### Boxplots of ClusterSum abundance vs PD and variants (p-value < 0.05)

Only a few associations are significant here. Below you can find all significant
ones in boxplots taking into consideration genotypes and PD case/control status.
Plots are sorted by decreasing absolute $\beta$ values.

```{r, fig.height=12, fig.width=8}
plot_top_assoc <- lm_results_interaction_bonferoni.df %>%
  filter(grepl("chr[0-9]", term), p.value < 0.05) %>% 
  arrange(desc(abs(estimate)))

ggarrange(plotlist = lapply(1:nrow(plot_top_assoc), function(i) {
  clst <- plot_top_assoc$clusterName[i]
  varnt <- str_extract(plot_top_assoc$term[i], "chr.*")
  bonfSignif <- plot_top_assoc$p.adj_bonferroni[i] < 0.1
  
  majAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.) - 1]
  minAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.)]
  tmp <- data_alltogether.df %>% set_names(make.names(colnames(.)))
  
  tmp[,varnt] <- case_when(
        round(tmp[[varnt]], 0) == 0 ~ paste(majAll, majAll, sep = "|"),
        round(tmp[[varnt]], 0) == 1 ~ paste(majAll, minAll, sep = "|"),
        round(tmp[[varnt]], 0) == 2 ~ paste(minAll, minAll, sep = "|"),
        ) %>% 
    factor(levels = c(paste(majAll, majAll, sep = "|"), paste(majAll, minAll, sep = "|"), paste(minAll, minAll, sep = "|")))
  
  return(ggplot(tmp,
         aes(
           x = eval(
             parse(text = varnt)
           ),
           y = eval(
             parse(text = make.names(clst)
           )),
           color = Case_status
         )) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2)) +
    geom_boxplot(alpha = 0.4,
                 show.legend = FALSE,
                 outliers = FALSE) +
    scale_color_manual(values = c("skyblue3", "gold3")) +
    labs(
      x = "Genotype Dosages (rounded)",
      y = "ClusterSum (crude)",
      title = paste(clst, "vs", varnt),
      color = "Disease status",
      caption = paste("Bonferroni significant:", bonfSignif)
    )
  )
  }), ncol = 2, nrow = 4,
  common.legend = TRUE)
```

## Fold Change PD/healthy split by genotype dosage (0, 1, 2)

Now we want to ask what is the fold change of a cluster (e.g. Fiber degraders)
between PD and Controls having the same genotype. We would expect some fold changes
to increase at increasing genotype, while adjusting for age and sequencing depth.

$$
ClusterSum_i \sim Age + Seq.Depth + PD
$$

And extract the Fold Change of each cluster for PD $FC_i = \frac{PD}{Ctrl}$. Here
crude p-values are of interest, so we will not adjust them.

```{r}
lm_split_by_variant_and_dosage <- data_alltogether.df %>%
  pivot_longer(
    cols = colnames(select(per_sample_dosages.df, -sample_name)), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = colnames(select(cluster_sums.df, -sample_name)), 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>%
  mutate(dosage = factor(round(dosage, 0))) %>%  # Convert dosage to a factor
  group_by(clusterName, variant, dosage) %>%  # Group by clusterName, variant, and dosage
  filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 5 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()

write_tsv(lm_split_by_variant_and_dosage, file = file.path(default_outdir, "02_lm_by_genotype.tsv"))
```

```{r table split results}
signif_assocs <- lm_split_by_variant_and_dosage %>%
  filter(term == "Case_statusPD", p.value < 0.1)
  
summary_table <- lm_split_by_variant_and_dosage %>% 
  filter(term == "Case_statusPD") %>% 
  mutate(
    estimate_ci = sprintf("%.3f (%.3f, %.3f)", 
                          estimate, 
                          estimate - 1.96 * std.error, 
                          estimate + 1.96 * std.error),
    p_adj = sprintf("%.3e",p.value)
  ) %>%
  select(clusterName, variant, dosage, estimate_ci, p_adj) %>% 
  pivot_wider(
    names_from = dosage, 
    values_from = c(estimate_ci, p_adj),
    names_glue = "{.value}_{dosage}"
  ) %>%
  arrange(clusterName, variant) %>% 
  .[,c(1,2,3,6,4,7,5,8)] %>% 
  ungroup() %>% 
  filter(as.numeric(p_adj_0) < 0.05 | as.numeric(p_adj_1) < 0.05 | as.numeric(p_adj_2) < 0.05) %>% 
  mutate(
    p_adj_0 = ifelse(as.numeric(p_adj_0) < 0.05, paste(p_adj_0, "*"), p_adj_0),
    p_adj_1 = ifelse(as.numeric(p_adj_1) < 0.05, paste(p_adj_1, "*"), p_adj_1),
    p_adj_2 = ifelse(as.numeric(p_adj_2) < 0.05, paste(p_adj_2, "*"), p_adj_2)
  )

highlight_rows <- which(
  as.numeric(gsub(" *", "", summary_table$p_adj_0, fixed = TRUE)) < 0.05 | 
    as.numeric(gsub(" *", "", summary_table$p_adj_1, fixed = TRUE)) < 0.05 |
    as.numeric(gsub(" *", "", summary_table$p_adj_2, fixed = TRUE)) < 0.05
  )


summary_table %>% 
  as.matrix() %>% 
  magrittr::set_colnames(c("clusterName", "variant", rep(c("Beta (95% C.I.)", "P.val"), 3))) %>%
  
  kable("html", caption = "Overview of clusters' associations with PD compared to Controls split by each variant's dosage. Only variants with at least 1 significant association in any of the 3 dosages were kept in the table. PD-cluster associations that passed Bonferroni test as shown above, were marked with an asterisk ('*') on the P.val column. 'Beta' values indicate the association between sex. The analysis was restricted to variants with at least 5 individuals with PD and 5 Controls in each dosage. NAs are cases where dosages did not have enough cases and controls to model.") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = T) %>%
  add_header_above(c(" " = 2, "Dosage 0" = 2, "Dosage 1" = 2, "Dosage 2" = 2)) %>%
  collapse_rows(columns = 1, valign = "top")
```

```{r}
lm_split_by_variant_and_dosage %>% 
  filter(term == "Case_statusPD") %>% 
  complete(clusterName, variant, dosage, fill = list(estimate = 0, std.error = 0, statistic = 0, p.value = 0)) %>% 
  ggplot(aes(x = variant, y = estimate, fill = dosage)) + 
  geom_bar(stat = "identity", position = position_dodge()) + 
  facet_grid(rows = vars(clusterName), cols = vars(variant), scale = "free_x") + 
  scale_fill_brewer() +
  theme(
    axis.text.x = element_blank(),
    legend.position = "top",
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90)
    ) + 
  labs(
    x = NULL
  )

```

The y axis in this case is the fold change between PD vs controls in each genotype.
Larger values mean more difference between PD and healthy patients.

  * `Bifidobacteria` show variations of fold change values at increasing dosage 
  in different fashions:
    
    - Quantitative downward for variant chr4_89587189
    - Quantitative upward for variant chr4_89587189
    - Recessive downward (high in homRef, low elsewhere): chr4_89685975, 
      chr4_89715479, chr4_89716450, chr4_89720189, chr4_89744890
    - Recessive upward: chr6_32602640, chr6_32604184
    
  * `Fiber Degraders` show similar and somewhat opposite trends:
  
    - Quantitative downward: chr4_89587189, chr4_89840793
    - Dominant downward: chr4_89828405, chr6_32441753
    
### The other way around: cluster vs dosage split by health status 

The following model was done in health and PD participants separately

$$
clusterRelAb \sim Age + Seq.Depth + dosage
$$

```{r}
lm_dosage_split_by_PD <- data_alltogether.df %>%
  pivot_longer(
    cols = colnames(select(per_sample_dosages.df, -sample_name)), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = colnames(select(cluster_sums.df, -sample_name)), 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>%
  group_by(clusterName, variant, Case_status) %>%  # Group by clusterName, variant, and dosage
  #filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 5 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + dosage, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()
```

```{r}

lm_dosage_split_by_PD %>% 
  filter(term == "dosage") %>% 
  ggplot(aes(x = variant, y = estimate, fill = Case_status)) + 
  geom_col(position = position_dodge()) + 
  facet_grid(rows = vars(clusterName), cols = vars(variant), scale = "free_x") + 
  scale_fill_manual(name = "Case Status", values = c("skyblue3", "gold3")) + 
  theme(
    axis.text.x = element_blank(),
    legend.position = "top",
    strip.text.y = element_text(angle = 0),
    strip.text.x = element_text(angle = 90)
    ) + 
  labs(
    x = NULL
  )

```

The plot above shows the effect sizes of the associations between bacterial 
clusters and variant dosage **split by healthy and PD subjects**. Discordant
effect sizes, although not significant in most cases, suggest an opposite dose
effect among PD and health carriers. 

  * Bifidos tend to decrease on PD at increasing dosage of several 
  variants on SNCA, but increase in other cases. In healthy participants, there
  tend to be opposite trends, or none

  * Fiber degraders tend to be inversely proportional to variants' dosages, 
  but only in PD cases. In healhy controls it is the opposite trend.

  * As usual, other bacterial clusters do not show noticeable trends
  
# Approach 2: Microbiome wide regression with another method, followed by GSEA (WIP)

As a sensitivity analysis, we can agnostically test all bacteria for association 
with genetic variants in interaction with PD. I have used `linda` for microbiome
modeling, one model for each variant. Differential abundance was followed by 
gene set enrichment analysis (GSEA) performed with standard parameters with the
`fgsea` Bioconductor package. As microbiome sets we can use:

  - the clusters generated by Haydeh
  - Other clusters as reported in `bugphyzz`, a repository of experimental data 
  on bacterial physiology. 
  
The second one is even more agnostic, but it likely misses some cases that 
Haydeh manually curated with literature search.

## LinDA on SNCA and HLA

The model done here is identical to the first approach:

$$
ClusterSum_i \sim Age + Seq.Depth + PD + Variant_k + PD \times Variant_k
$$

And we extract coefficients (logFC) in relation to **dosage only** and 
**dosage in PD** to perform GSEA using Haydeh's clusters

```{r}
samples_to_keep.df <- colData(DoD_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() %>% 
  filter(collection_method != "swab")

variants_to_keep_rowData.df <- as.data.frame(rowData(DoD_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(DoD_data.mae[["genotypes_dosages.se"]])$GENE),])

per_sample_dosages.df <- t(assay(DoD_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("sample_name")

meta_with_variants <- inner_join(per_sample_dosages.df, samples_to_keep.df, by = "sample_name")

```

```{r}
if(!("MicrobiomeStat" %in% installed.packages())){
  install.packages("MicrobiomeStat")
  }

library(MicrobiomeStat)

genotypes_vs_microbiome_linda.list <- sapply(grep("^chr", colnames(meta_with_variants), value = TRUE), 
       function(gt)
       suppressMessages(linda(
         feature.dat = assay(DoD_data.mae, "metaphlan_Counts.se")[,meta_with_variants$sample_name],
         meta.dat = meta_with_variants,
  formula = paste("~ Age_at_collection + seqs_scaled + Case_status +", gt, "+", paste(gt, "Case_status", sep = "*")), 
  p.adj.method = "BH",
  corr.cut = 0.1,
  outlier.pct = 0.03, 
  alpha = 0.1,
  prev.filter = 0.05,
  verbose = FALSE
  )),
  USE.NAMES = TRUE,
  simplify = FALSE
  )

linda_results_wrangled_variants_SNCA_HLA.df <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply(function(x) x[grepl("chr", names(x))] %>% set_names(c("variantOnly", "interactionPD")) %>% lapply(rownames_to_column, "taxon")) %>% 
  lapply(bind_rows, .id = "covariate") %>% 
  bind_rows(.id = "variantName")

write_tsv(linda_results_wrangled_variants_SNCA_HLA.df, file = file.path(default_outdir, "LinDA_regression_FullResults.tsv"))
```

## GSEA on variant-only effects {.tabset}

These results confirm our findings above, showing the tendency of enrichment of 
Fiber Degraders in case of dosage only effect, while for the same variant, there 
is an opposite trend. 

This is particularly visible in variant `rs983361`, named chr4_89840793

```{r}
# fgsea on variant only effects
variant_only_t_stats <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 4) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variantsonly <- lapply(variant_only_t_stats, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats))

gsea_on_variantsonly_someSignif <- gsea_on_variantsonly[sapply(gsea_on_variantsonly, function(x) any(x$padj < 0.05))]
```

```{r, results="asis"}
for (i in names(gsea_on_variantsonly_someSignif)) {
    cat("\n###", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_only_t_stats[[i]],
      fgseaRes = gsea_on_variantsonly_someSignif[[i]])
      )
    cat("\n")
}
```

## {-}

## GSEA on variant-PD interactions {.tabset}

```{r}
# fgsea on variant only effects
variant_PD_t_stat <- lapply(genotypes_vs_microbiome_linda.list, "[[", "output") %>% 
  lapply("[[", 5) %>% 
  lapply(function(x) set_names(x$stat, rownames(x)))

library(fgsea)

gsea_on_variant_PD <- lapply(variant_PD_t_stat, function(stats) 
                               fgsea(
  pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name),
  stats = stats))

gsea_on_variant_PD_someSignif <- gsea_on_variant_PD[sapply(gsea_on_variant_PD, function(x) any(x$padj < 0.05))]
```

```{r, results="asis"}
for (i in names(gsea_on_variant_PD_someSignif)) {
    cat("\n###", i, sep = " ")
      
    cat("\n")
    
    variants_to_keep_rowData.df %>%
      filter(POS == strsplit(i, "_")[[1]][2]) %>%
      select(CHROM, POS, REF, ALT, RSID, GENE, IMPUTED = INFO_IMPUTED, R2 =  INFO_R2) %>%
      kbl() %>%
      print()
    
    cat("\n")
    
    print(
      plotGseaTable(
      pathways = split(biomarkers_medrxiv.df$species_metaphlan, biomarkers_medrxiv.df$cluster_name), 
      stats = variant_PD_t_stat[[i]],
      fgseaRes = gsea_on_variant_PD_someSignif[[i]])
      )
    cat("\n")
}
```

## {-}

# Approach 2.5 - GSEA on attributes found in `bugphyzz`

This approach is even more unbiased. The analysis is based on their vignette.
(https://www.bioconductor.org/packages/release/data/experiment/vignettes/bugphyzz/inst/doc/bugphyzz.html)

```{r}
# BiocManager::install("bugphyzz") 
library(bugphyzz)
bp <- importBugphyzz()

names(bp)
```

There are many `data.frame`s with bug classification, with some overlap expected.
These sets can be used to test enrichment of a particular trait. In particular,
aerophilicity, gram staining, pathogenicity and butyrate production could be 
of interest.

Mapping can be done using `Taxon_name` and `Rank`

```{r}
bp$`butyrate-producing bacteria`

```

As can be seen, the list is not too large, in this case, and mapping is hard.