---
title: "PD, microbiome and host genetics on SNCA and HLA variants"
subtitle: "Data: UAB"
author: "Giacomo Antonello, Zachary D. Wallen, Charles M. Murchinson, Levi Waldron, Haydeh Payami"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
---

```{r, echo=FALSE}
knitr::opts_chunk$set(
  message = FALSE, 
  warning = FALSE
)

default_outdir <- "output_UAB/"
dir.create(default_outdir, showWarnings = FALSE)
```

```{r, include=FALSE}
# Check and install necessary CRAN packages
CRAN_required_packages <- c("tidyverse", "gwasrapidd", "DT", "kableExtra", "data.table", "ggpubr", "BiocManager", "readxl","kableExtra")

install.packages(
  CRAN_required_packages[!(CRAN_required_packages %in% installed.packages())] 
)

# Check and install necessary Bioconductor packages
BioC_required_packages <- c("Maaslin2", "ANCOMBC", "biomaRt", "mia", "phyloseq", "microbiome","SIAMCAT", "fgsea")
BioC_required_packages <- BioC_required_packages[!(BioC_required_packages %in% installed.packages())]

if(length(BioC_required_packages) > 0) {
  BiocManager::install(BioC_required_packages[!(BioC_required_packages %in% installed.packages())],
                       ask = FALSE,
                       update = TRUE)
}


# load basic data manipulation packages
library(phyloseq)
library(DT)
library(kableExtra)
library(MultiAssayExperiment)
library(ggpubr)
library(tidyverse)

theme_set(theme_light())

rm(BioC_required_packages,CRAN_required_packages)
gc(verbose = FALSE)
```

Below is the data sources for UAB. This has been 
wrangled by Giacomo to make one single data object that would, hopefully, speed
things up in future analyses. This data object contains microbiome, pathway, 
participants data and genetic data.

```{r}
# load previously wrangled data
load("../Data/UAB_data.mae.RData")

UAB_data.mae
```

Genetic data have been retrieved with the protocol available in `00_UAB_DataPreparation.html`.
In brief:

  1. The GWAS catalog was searched for variants associated with the ontology term
  `MONDO_0005180`, and downloaded. all this with 
  [gwasrapidd](https://rmagno.eu/gwasrapidd/articles/gwasrapidd.html). 
  Catalog's version is not stated, I'm assuming that since the packages uses the `REST API`, 
  it should be the newest.
  
  2. Gene annotation was taken from `biomaRt`, but the GWAS catalog had annotations
  too. Naming is based on the HUGO nomenclature.

# Assocation between PD markers and SNCA and HLA gene variants  (ClustSum)

## Define `ClusterSum`

Haydeh has manually curated a list of bacteria that go together into 6 different
clusters elevated or decreased in PD. To see the genetic impact of PD associated
variants on these bacteria without having to pay a high multiple testing price.

Our first approach is to do the following:

  1. Define which species belong to which cluster
  2. Divide % abundances by 100, to obtain relative abundances again
  3. Sum relative abundances of all species within each cluster

```{r}
biomarkers_medrxiv.df <- readxl::read_xlsx("../Data/Biomarkers_Payami_medRxiv.xlsx")

datatable(biomarkers_medrxiv.df %>% select(-species_metaphlan, `Number of connections`))
```

```{r}

cluster_sums.df <- left_join(biomarkers_medrxiv.df, 
          as.data.frame(assay(UAB_data.mae, "metaphlan_RelAbund.se")/100) %>%
            rownames_to_column("species_metaphlan"), 
          by = "species_metaphlan") %>%
  split.data.frame(f = .$cluster_name) %>% 
  lapply(select, all_of(MultiAssayExperiment::colnames(UAB_data.mae)[["metaphlan_RelAbund.se"]])) %>% 
  lapply(colSums) %>% 
  bind_cols(SubjectID = names(.[[1]])) %>% 
  relocate(SubjectID)

cluster_sums.df %>% 
  mutate_if(is.numeric, round, 3) %>% 
datatable(caption = "Examples of values obtained with the sum method across clusters. These will be the variables we will use to test association with variant dosages")
```

These variables are far from normal, and log transformation does not help.

```{r, fig.cap= "Distribution of cumulative sum of relative abundances of cluster member species."}
cluster_sums.df %>% 
  select(-SubjectID) %>% 
  reshape2::melt() %>% 
  mutate(
    log2Transf = log2(value + 1),
    sq.root = sqrt(value),
    clr = compositions::clr(value + 1)
    ) %>% 
  reshape2::melt(
    variable.name = "Transformation"
    ) %>% 
  mutate(
    Transformation = case_when(
      Transformation == "value" ~ 'untransformed',
      Transformation == "log2Transf" ~ 'log2',
      Transformation == "sq.root" ~ 'sq.root',
      Transformation == "clr" ~ 'clr',
      TRUE ~ 'error'
      )
    ) %>% 
  ggplot(aes(x = value, color = Transformation)) + 
  geom_histogram(fill = "transparent") + 
  facet_wrap(~variable) 
```

## Define variants of interest and number of independent tests

Genetic variants were included if retained if:
  - they were associated with PD in any GWAS catalog entry
  - they were found in the Genetic data
  - they had a gene name associated to them (for easy lookup)


```{r}

samples_to_keep.df <- colData(UAB_data.mae[["metaphlan_RelAbund.se"]]) %>%
  as.data.frame() %>% 
  filter(
    !is.na(Age_at_collection),
    !is.na(seqs_scaled),
    !is.na(Case_status)
    )

variants_to_keep_rowData.df <- as.data.frame(rowData(UAB_data.mae[["genotypes_dosages.se"]])[grepl("HLA|SNCA", rowData(UAB_data.mae[["genotypes_dosages.se"]])$GENE),])
rownames(variants_to_keep_rowData.df) <- variants_to_keep_rowData.df$RSID

per_sample_dosages.df <- t(assay(UAB_data.mae, "genotypes_dosages.se"))[,rownames(variants_to_keep_rowData.df)] %>% 
  as.data.frame() %>% 
  set_names(paste(variants_to_keep_rowData.df$CHROM, variants_to_keep_rowData.df$POS, variants_to_keep_rowData.df$REF, variants_to_keep_rowData.df$ALT, sep = "_")) %>% 
  rownames_to_column("SubjectID")

data_alltogether.df <- inner_join(cluster_sums.df, per_sample_dosages.df, by = "SubjectID") %>% 
  inner_join(samples_to_keep.df, by = "SubjectID")

as_tibble(variants_to_keep_rowData.df) %>% 
  set_names(
    gsub("INFO_", "", colnames(.))
    ) %>%
  rename(
    `Empirical (R2)` = ER2,
    `Imputation Quality (R2)` = R2,
         ) %>% 
  datatable(caption = "List of variants found in the GWAS catalog that are explicitly mapped to the SNCA or HLA genes.")
```

Some of these variants are in high LD. We want to adjust for the total number of 
independent tests, therefore we adjust for variants in low LD only across the genes
of interest. For the sake of reproducibility, we will use a `LDlink` (by NIH) 
API to prune the variants of interest.

```{r LDlink parameters setup}
r2_thresh <- 0.1
maf_thresh <- 0.01
```

```{r LD pruning}

cache(
  variants_pruning_step <- split.data.frame(variants_to_keep_rowData.df,variants_to_keep_rowData.df$CHROM) %>% 
  lapply(
    function(x)
    LDlinkR::SNPclip(
      r2_threshold = r2_thresh,
      snps = x$RSID,
      pop = "CEU",
      maf_threshold = maf_thresh,
      token = "78e50d5fd7e7", 
      genome_build = "grch38"
      )
  ), dir = file.path(default_outdir, "variant_pruning_cache")
  )

variants_pruning_table <- tryCatch(lapply(
  variants_pruning_step, mutate, Details = ifelse(grepl("removed", Details), "Removed", "Kept")
  ) %>% 
  lapply(group_by, Details) %>%
  lapply(tally) %>%
  bind_rows(.id = "Chromosome") %>% 
  pivot_wider(names_from = "Details", values_from = "n") %>% 
  mutate(Total = Kept + Removed),
  error = function(e) return(NA))

kbl(variants_pruning_table, caption = paste0("Number of idependent variants (Kept) based on LDlink calculations on default parameters (R2 = ", r2_thresh, ", MAF = ", maf_thresh, "). The population chosen is 'CEU'. The sum of 'Kept' values represents the number of independent tests that need to be accounted for later on."))
```

```{r define number of tests to account for}
n_indep_variants <- ifelse(all(is.na(variants_pruning_table)), 6, sum(sapply(variants_pruning_step, function(x) sum(grepl("kept", x$Details)))))
```

## Sample inclusion criteria and statistical modeling

  - Collection method should not be `swab` 
  - Available data for age, sequencing depth, microbiome PD cumulative scores, genetic data

This results in `r nrow(data_alltogether.df)` samples for the analysis, with variant's distributions as follows:

```{r}
cbind(data_alltogether.df %>% 
  select(starts_with("chr")) %>% 
  mutate_all(round, 0) %>% 
  sapply(table) %>% 
  t(), 
  data_alltogether.df %>% 
    select(starts_with("chr")) %>% 
    mutate_all(round, 0) %>% 
    sapply(function(x) table(x) %>% prop.table()) %>% 
    t()) %>% 
  .[,c(1,4,2,5,3,6)] %>%   
  as.data.frame() %>% 
  set_names(make.unique(colnames(.))) %>% 
  rownames_to_column("id") %>%
  separate_wider_delim(cols = id, names = c("CHROM", "POS", "REF", "ALT"), delim = "_", too_few = "align_start") %>% 
  rename(
    `0 (N)` = `0`,
    `0 (%)` = `0.1`,
    `1 (N)` = `1`,
    `1 (%)` = `1.1`,
    `2 (N)` = `2`,
    `2 (%)` = `2.1`
  ) %>% 
  mutate_at(.vars = vars(c(`0 (%)`,`1 (%)`,`2 (%)`)), .funs = function(x) round(x*100, 2)) %>% 
  
  kbl(caption = "Number of samples with alternate allele dosages. NB: no distinction is done for 1|0 or 0|1, as both carry one alternate allese. Values are shown as total counts and percentages") %>% 
  kable_styling() %>% 
  add_header_above(c("Variant Information" = 4, "Distribution of dosages in the cohort studied" = 6)) %>% 
  column_spec(c(4,6,8), border_right = TRUE)
```

Now perform the model with an interaction term between PD and each variant.

$$
ClusterSum_i \sim Age + Seq.Depth + PD + Variant_k + PD \times Variant_k
$$

```{r}

lm_results_interaction.df <- data_alltogether.df %>%
  pivot_longer(
    cols = starts_with("chr"), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = colnames(select(cluster_sums.df, -SubjectID)), 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>% 
  group_by(clusterName, variant) %>%  # Group by clusterName, variant, and dosage
#  filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 5 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + Case_status + dosage + dosage*Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()


# number of tests to correct for as bonferroni: number of cluster + variants tested
n_tests <- length(unique(lm_results_interaction.df$clusterName)) * n_indep_variants

lm_results_interaction_bonferoni.df <- lm_results_interaction.df %>% 
  mutate(
    p.adj_bonferroni = p.value * n_tests,
    p.adj_bonferroni = ifelse(p.adj_bonferroni >1, 1, p.adj_bonferroni),
    significant = ifelse(p.adj_bonferroni < 0.1, TRUE, FALSE)
  )

write_tsv(lm_results_interaction_bonferoni.df, file = file.path(default_outdir, "01_lm_interaction_Variant.x.PD.tsv"))

```

We correct p-values using the Bonferroni method accounting for `r 6` clusters tested
against `r n_indep_variants` independent variants as defined above, for a total of `r n_tests` 
independent tests. Then, we consider associations between clusters and variants 
with `Adj. P-value < 0.05` as significant further annotating whether each association 
passed family-wise error rate adjustment.

```{r bugs and genetics results table}
lm_results_interaction_bonferoni.df %>%
  filter(grepl("dosage", term)) %>% 
  arrange(desc(abs(estimate))) %>% 
  select(-statistic) %>% 
  mutate_if(is.numeric, round, 4) %>% 
  reactable::reactable(filterable = TRUE, searchable = TRUE, sortable = TRUE, columns = list(variant = reactable::colDef(width = 200), term = reactable::colDef(width = 200)))
```

## Visualization of variant-cluster associations with P-value < 0.05

### Betas bar plots

```{r}
vgsub <- Vectorize(gsub)

lm_results_interaction_bonferoni.df <- mutate(lm_results_interaction_bonferoni.df, term = vgsub("dosage", variant, term)) %>% 
  select(-variant)

data_plot.tmp <- lm_results_interaction_bonferoni.df %>% 
  filter(grepl("chr", term)) %>%
  mutate(
    variant = str_extract(term, "chr.*"),
    PD = ifelse(grepl("PD", term), "PD", "Control"), 
    term = NULL) %>% 
  # used for older plot version, but good to keep for future uses
  mutate(
    color_manual = case_when(
      p.adj_bonferroni < 0.1 & PD == "Control" ~ "skyblue3",
      p.adj_bonferroni < 0.1 & PD == "PD" ~ "gold3",
      p.adj_bonferroni >= 0.1 ~ "gray"
  )
  )

plotly::ggplotly(
ggplot() +
  geom_col(data = data_plot.tmp, mapping = aes(x = variant, y = estimate, fill = PD)) +
  geom_text(data = data_plot.tmp %>% filter(p.value < 0.05, p.adj_bonferroni >= 0.1), mapping = aes(x = variant, y = estimate, label = "*")) +
  geom_text(data = data_plot.tmp %>% filter(p.adj_bonferroni < 0.1), mapping = aes(x = variant, y = estimate, label = "**")) +
  scale_fill_manual(values = c("gold3", "skyblue3")) +
  facet_wrap(~ clusterName) +
  geom_hline(yintercept = 0, color = "gray30") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(
    y = "Beta estimates", 
    title = "Full Betas range",
    caption = "*: Nominal p-value < 0.05, **: Bonferroni-adjusted P-value < 0.1"
  ), 
)
```

### Boxplots of ClusterSum vs PD and variants (p-value < 0.05)

Only a few associations are significant here. Below you can find all significant
ones in boxplots taking into consideration genotypes and PD case/control status.
Plots are sorted by decreasing absolute $\beta$ values.

```{r, fig.height=12, fig.width=8}
plot_top_assoc <- lm_results_interaction_bonferoni.df %>%
  filter(grepl("chr[0-9]", term), p.value < 0.05) %>% 
  arrange(desc(abs(estimate)))

ggarrange(plotlist = lapply(1:nrow(plot_top_assoc), function(i) {
  clst <- plot_top_assoc$clusterName[i]
  varnt <- str_extract(plot_top_assoc$term[i], "chr.*")
  bonfSignif <- plot_top_assoc$p.adj_bonferroni[i] < 0.1
  
  majAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.) - 1]
  minAll <- str_split(varnt, "_") %>% .[[1]] %>% .[length(.)]
  tmp <- data_alltogether.df %>% set_names(make.names(colnames(.)))
  
  tmp[,varnt] <- case_when(
        round(tmp[[varnt]], 0) == 0 ~ paste(majAll, majAll, sep = "|"),
        round(tmp[[varnt]], 0) == 1 ~ paste(majAll, minAll, sep = "|"),
        round(tmp[[varnt]], 0) == 2 ~ paste(minAll, minAll, sep = "|"),
        ) %>% 
    factor(levels = c(paste(majAll, majAll, sep = "|"), paste(majAll, minAll, sep = "|"), paste(minAll, minAll, sep = "|")))
  
  return(ggplot(tmp,
         aes(
           x = eval(
             parse(text = varnt)
           ),
           y = eval(
             parse(text = make.names(clst)
           )),
           color = Case_status
         )) +
    geom_point(position = position_jitterdodge(jitter.width = 0.2)) +
    geom_boxplot(alpha = 0.4,
                 show.legend = FALSE,
                 outliers = FALSE) +
    scale_color_manual(values = c("skyblue3", "gold3")) +
    labs(
      x = "Genotype Dosages (rounded)",
      y = "ClusterSum (crude)",
      title = paste(clst, "vs", varnt),
      color = "Disease status",
      caption = paste("Bonferroni significant:", bonfSignif)
    )
  )
  }), ncol = 2, nrow = 4,
  common.legend = TRUE)
```

## Fold Change PD/healthy split by genotype dosage (0, 1, 2)

Now we want to ask what is the fold change of a cluster (e.g. Fiber degraders)
between PD and Controls having the same genotype. We would expect some fold changes
to increase at increasing genotype, while adjusting for age and sequencing depth.

$$
ClusterSum_i \sim Age + Seq.Depth + PD
$$

And extract the Fold Change of each cluster for PD $FC_i = \frac{PD}{Ctrl}$. Here
crude p-values are of interest, so we will not adjust them.

```{r}
lm_split_by_variant_and_dosage <- data_alltogether.df %>%
  pivot_longer(
    cols = colnames(select(per_sample_dosages.df, -SubjectID)), 
    names_to = "variant", 
    values_to = "dosage"
  ) %>%
  pivot_longer(
    cols = colnames(select(cluster_sums.df, -SubjectID)), 
    names_to = "clusterName",
    values_to = "clusterRelAb"
  ) %>%
  mutate(dosage = factor(round(dosage, 0))) %>%  # Convert dosage to a factor
  group_by(clusterName, variant, dosage) %>%  # Group by clusterName, variant, and dosage
  filter(sum(Case_status == "PD") >= 5 & sum(Case_status == "Control") >= 5) %>%  # Keep only groups with at least 2 observations
  group_modify(~ {
    # Apply linear model for each group and tidy the output
    lm_results <- lm(clusterRelAb ~ Age_at_collection + seqs_scaled + Case_status, data = .x)
    broom::tidy(lm_results)
  }) %>% 
  ungroup()
```

```{r table split results}
signif_assocs <- lm_split_by_variant_and_dosage %>%
  filter(term == "Case_statusPD", p.value < 0.1)
  
summary_table <- lm_split_by_variant_and_dosage %>% 
  filter(term == "Case_statusPD") %>% 
  mutate(
    estimate_ci = sprintf("%.3f (%.3f, %.3f)", 
                          estimate, 
                          estimate - 1.96 * std.error, 
                          estimate + 1.96 * std.error),
    p_adj = sprintf("%.3e",p.value)
  ) %>%
  select(clusterName, variant, dosage, estimate_ci, p_adj) %>% 
  pivot_wider(
    names_from = dosage, 
    values_from = c(estimate_ci, p_adj),
    names_glue = "{.value}_{dosage}"
  ) %>%
  arrange(clusterName, variant) %>% 
  .[,c(1,2,3,6,4,7,5,8)] %>% 
  ungroup() %>% 
  filter(as.numeric(p_adj_0) < 0.05 | as.numeric(p_adj_1) < 0.05 | as.numeric(p_adj_2) < 0.05) %>% 
  mutate(
    p_adj_0 = ifelse(as.numeric(p_adj_0) < 0.05, paste(p_adj_0, "*"), p_adj_0),
    p_adj_1 = ifelse(as.numeric(p_adj_1) < 0.05, paste(p_adj_1, "*"), p_adj_1),
    p_adj_2 = ifelse(as.numeric(p_adj_2) < 0.05, paste(p_adj_2, "*"), p_adj_2)
  )

highlight_rows <- which(
  as.numeric(gsub(" *", "", summary_table$p_adj_0, fixed = TRUE)) < 0.05 | 
    as.numeric(gsub(" *", "", summary_table$p_adj_1, fixed = TRUE)) < 0.05 |
    as.numeric(gsub(" *", "", summary_table$p_adj_2, fixed = TRUE)) < 0.05
  )


summary_table %>% 
  as.matrix() %>% 
  magrittr::set_colnames(c("clusterName", "variant", rep(c("Beta (95% C.I.)", "P.val"), 3))) %>%
  
  kable("html", caption = "Overview of clusters' associations with PD compared to Controls split by each variant's dosage. Only variants with at least 1 significant association in any of the 3 dosages were kept in the table. PD-cluster associations that passed Bonferroni test as shown above, were marked with an asterisk ('*') on the P.val column. 'Beta' values indicate the association between sex. The analysis was restricted to variants with at least 5 individuals with PD and 5 Controls in each dosage. NAs are cases where dosages did not have enough cases and controls to model.") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = T) %>%
  add_header_above(c(" " = 2, "Dosage 0" = 2, "Dosage 1" = 2, "Dosage 2" = 2)) %>%
  collapse_rows(columns = 1, valign = "top")
```

```{r}
lm_split_by_variant_and_dosage %>% 
  filter(term == "Case_statusPD") %>% 
  complete(clusterName, variant, dosage, fill = list(estimate = 0)) %>% 
  ggplot(aes(x = variant, y = estimate, fill = dosage)) + 
  geom_bar(stat = "identity", position = position_dodge()) + 
  facet_wrap(~clusterName)+
  scale_fill_brewer() +
  geom_text(
    data = lm_split_by_variant_and_dosage %>%
      filter(term == "Case_statusPD", p.value < 0.5) %>%
      complete(clusterName, variant, dosage, fill = list(estimate = 0)) %>%
      mutate(yposit = sign(estimate)*(abs(estimate) + 0.01)), 
    mapping = aes(x = variant, y = yposit, label = ifelse(p.value < 0.05, "*", ""), group = dosage), 
            position = position_dodge(width = 1), size = 4) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), legend.position = "top") +
  labs(
    x = "Variant",
    y = expression(Eff.~Size~(beta)),
    fill = "Dosage"
  )


write_tsv(lm_split_by_variant_and_dosage, file = file.path(default_outdir, "02_lm_by_genotype.tsv"))

```

# Dirichlet Multinomial clustering correlates with Haydeh's picks?

This is a Bayesian approach of clustering counts data, works pretty nicely, 
assuming compositionality.

```{r DirichletMultinomial calculation, eval=interactive()}
library(DirichletMultinomial)

counts_tab <- t(assay(UAB_data.mae, "metaphlan_Counts.se"))

#' model up to 10 allowed clusters
dmn_raw_output <- parallel::mclapply(
  1:10 %>% set_names(paste("k =", .)),
  function(n)
    dmn(count = counts_tab, k = n, verbose = interactive(), seed = 2993), 
  mc.cores = 8
  )

dmn_outdir <- file.path(default_outdir, "DirichletMultinomial")
dir.create(dmn_outdir, showWarnings = FALSE)
saveRDS(dmn_raw_output, file = paste0(dmn_outdir, "/DMN_raw_output_", Sys.Date(), ".Rds"))
```

```{r}
k <- 6
```

Dirichlet components will be forced to have `r k` clusters. Although the 
algorithm suggests 3 (The smaller the `laplace` estimate, the better the model).

```{r, eval=interactive()}

laplace_values <- sapply(dmn_raw_output, BIC)
plot(laplace_values, type = "b", col = "darkgray", xlab = "N. Dirichlet Components modeled", ylab = "BIC Fit evaluation")
points(x = which.min(laplace_values), y = laplace_values[which.min(laplace_values)] , pch = 16, cex = 1.5)

```

```{r, eval=interactive()}
#' find number of clusters with the least information (minimum of BIC curve)
clusters_arbitraryN <- dmn_raw_output[[k]]

CST_probs <- DirichletMultinomial::mixture(clusters_arbitraryN)

CST_assigns <- as.character(apply(CST_probs, 1, function(x) {
  best_x <- max(x)
  second_best_x <- max(x[x != best_x])
  CST_final <- ifelse(
    best_x < 0.9 & (best_x - second_best_x) < 0.1, 
    yes = paste(which(x == best_x), which(x == second_best_x), sep = "/"),
    no = which(x == best_x)
  )
  
  return(CST_final)
}))

CST_final <- cbind.data.frame(CST_assigns, CST_probs) %>% 
  set_names(c("CST_assign", paste("CST", 1:ncol(CST_probs), "prob", sep = "_"))) %>% 
  rownames_to_column("SubjectID")

metadata_with_CST.df <- inner_join(CST_final, data_alltogether.df, by = "SubjectID")

metadata_with_CST.df
```


```{r, eval=interactive()}
ggarrange(
  # plot fiber degraders
ggplot(metadata_with_CST.df, aes(x = Case_status, y = `Fiber Degraders`, color = CST_assign)) + 
  geom_point(position = position_jitterdodge(jitter.width = 0.1, seed = 1234)) +
  geom_boxplot(outliers = FALSE, alpha = 0.4) ,
  # plot Bifidobacteria
ggplot(metadata_with_CST.df, aes(x = Case_status, y = Bifidobacteria, color = CST_assign)) + 
  geom_point(position = position_jitterdodge(jitter.width = 0.1, seed = 1234)) +
  geom_boxplot(outliers = FALSE, alpha = 0.4),

ncol = 2, common.legend = TRUE
) 


```

As a quick look, Dirichlet components do not correlate well with clusters chosen
by Haydeh.

# Open Questions

  - Do Fiber degraders correlate with fiber-rich diets?
  - Do Bifidos correlate with lactose rich diets?
  - Do Bifidos correlate with the variant in LCT gene? Do these people have it at 
  all in the VCFs?
  
# R session

```{r}
sessionInfo()
```
